<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>analysis pipeline: from alignment to dated phyloscanner output</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore"><strong>analysis pipeline: from alignment to dated phyloscanner output</strong></h1>



<div id="input-data" class="section level2">
<h2><strong>Input data</strong></h2>
<p>The input data are sequence alignments that contain KC sequences and LANL sequences, for all subtypes.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">require</span>(tidyverse)</a>
<a class="sourceLine" id="cb1-2" title="2"></a>
<a class="sourceLine" id="cb1-3" title="3">home &lt;-<span class="st"> &#39;~/Box Sync/OR_Work/Seattle/analysis_191017&#39;</span></a>
<a class="sourceLine" id="cb1-4" title="4">indir &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;alignments&#39;</span>)   </a>
<a class="sourceLine" id="cb1-5" title="5"><span class="kw">list.files</span>(indir, <span class="dt">pattern=</span><span class="st">&#39;mafft</span><span class="ch">\\</span><span class="st">.fasta&#39;</span>)</a>
<a class="sourceLine" id="cb1-6" title="6"><span class="co">#&gt; [1] &quot;180709_LANL_Subtype01AE_mafft.fasta&quot;</span></a>
<a class="sourceLine" id="cb1-7" title="7"><span class="co">#&gt; [2] &quot;180709_LANL_SubtypeA1_mafft.fasta&quot;  </span></a>
<a class="sourceLine" id="cb1-8" title="8"><span class="co">#&gt; [3] &quot;180709_LANL_SubtypeB_mafft.fasta&quot;   </span></a>
<a class="sourceLine" id="cb1-9" title="9"><span class="co">#&gt; [4] &quot;180709_LANL_SubtypeC_mafft.fasta&quot;</span></a></code></pre></div>
</div>
<div id="step-1-remove-drug-resistance-mutations" class="section level2">
<h2><strong>Step 1 (remove drug resistance mutations)</strong></h2>
<p>I first mask in the alignments codons that are listed as major drug resistance mutations in the Stanford DB, and a list of resistance mutations that I got from Tulio.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1">seattle.<span class="fl">191017.</span>alignment.rm.drug.resistance.mutations &lt;-<span class="st"> </span><span class="cf">function</span>()</a>
<a class="sourceLine" id="cb2-2" title="2">{   </a>
<a class="sourceLine" id="cb2-3" title="3">    <span class="kw">require</span>(big.phylo)</a>
<a class="sourceLine" id="cb2-4" title="4">    home &lt;-<span class="st"> &#39;~/Box Sync/OR_Work/Seattle/analysis_191017&#39;</span></a>
<a class="sourceLine" id="cb2-5" title="5">    indir &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="st">&#39;alignments&#39;</span>)    </a>
<a class="sourceLine" id="cb2-6" title="6">    df &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">F=</span><span class="kw">list.files</span>(indir, <span class="dt">pattern=</span><span class="st">&#39;fasta$&#39;</span>,<span class="dt">full.names=</span><span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb2-7" title="7">    df &lt;-<span class="st"> </span><span class="kw">subset</span>(df, <span class="op">!</span><span class="kw">grepl</span>(<span class="st">&#39;nodrm&#39;</span>,F))</a>
<a class="sourceLine" id="cb2-8" title="8">    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_len</span>(<span class="kw">nrow</span>(df)))</a>
<a class="sourceLine" id="cb2-9" title="9">    {</a>
<a class="sourceLine" id="cb2-10" title="10">        infile.fasta &lt;-<span class="st"> </span>df[i,F]</a>
<a class="sourceLine" id="cb2-11" title="11">        outfile &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">.fasta&#39;</span>,<span class="st">&#39;_ndrm.fasta&#39;</span>,infile.fasta)</a>
<a class="sourceLine" id="cb2-12" title="12">        </a>
<a class="sourceLine" id="cb2-13" title="13">        <span class="kw">cat</span>(<span class="st">&#39;</span><span class="ch">\n</span><span class="st">process &#39;</span>,infile.fasta)</a>
<a class="sourceLine" id="cb2-14" title="14">        sq                  &lt;-<span class="st"> </span><span class="kw">read.dna</span>(infile.fasta, <span class="dt">format=</span><span class="st">&#39;fa&#39;</span>)  </a>
<a class="sourceLine" id="cb2-15" title="15">        tmp                 &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">grepl</span>(<span class="st">&quot;HXB2&quot;</span>,<span class="kw">rownames</span>(sq)))</a>
<a class="sourceLine" id="cb2-16" title="16">        <span class="kw">rownames</span>(sq)[tmp]   &lt;-<span class="st"> &#39;HXB2&#39;</span></a>
<a class="sourceLine" id="cb2-17" title="17">        tmp                 &lt;-<span class="st"> </span>big.phylo<span class="op">:::</span><span class="kw">seq.rm.drugresistance</span>(sq)</a>
<a class="sourceLine" id="cb2-18" title="18">        sq                  &lt;-<span class="st"> </span>tmp<span class="op">$</span>nodr.seq</a>
<a class="sourceLine" id="cb2-19" title="19">        <span class="kw">cat</span>(<span class="st">&#39;</span><span class="ch">\n</span><span class="st">write to &#39;</span>,outfile)</a>
<a class="sourceLine" id="cb2-20" title="20">        <span class="kw">write.dna</span>(sq, <span class="dt">file=</span> outfile, <span class="dt">format=</span><span class="st">&#39;fasta&#39;</span>)</a>
<a class="sourceLine" id="cb2-21" title="21">    }   </a>
<a class="sourceLine" id="cb2-22" title="22">}</a></code></pre></div>
</div>
<div id="step-2-make-bootstrap-alignments" class="section level2">
<h2><strong>Step 2 (make bootstrap alignments)</strong></h2>
<p>After the codons that may be drug resistance mutations are gone, I make bootstrap alignments. For now, I only made 10 though we may want to increase this later to 1000. I would like to use the bootstrap trees to find all possible groups of closely related individuals, and I am bit concerned that we may miss them in just one trees due to uncertainty in tree reconstruction.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">seattle.<span class="fl">191017.</span>alignment.make.bootstraps &lt;-<span class="st"> </span><span class="cf">function</span>()</a>
<a class="sourceLine" id="cb3-2" title="2">{   </a>
<a class="sourceLine" id="cb3-3" title="3">    home &lt;-<span class="st"> &#39;~/Box Sync/OR_Work/Seattle/analysis_191017&#39;</span></a>
<a class="sourceLine" id="cb3-4" title="4">    indir &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;alignments&#39;</span>)</a>
<a class="sourceLine" id="cb3-5" title="5">    outdir &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;alignments_bs&#39;</span>)</a>
<a class="sourceLine" id="cb3-6" title="6">    bsn &lt;-<span class="st"> </span><span class="dv">10</span></a>
<a class="sourceLine" id="cb3-7" title="7">    seed &lt;-<span class="st"> </span>42L</a>
<a class="sourceLine" id="cb3-8" title="8">    <span class="co">#</span></a>
<a class="sourceLine" id="cb3-9" title="9">    df &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">F=</span><span class="kw">list.files</span>(indir, <span class="dt">pattern=</span><span class="st">&#39;fasta$&#39;</span>,<span class="dt">full.names=</span><span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb3-10" title="10">    df &lt;-<span class="st"> </span><span class="kw">subset</span>(df, <span class="kw">grepl</span>(<span class="st">&#39;ndrm&#39;</span>,F))</a>
<a class="sourceLine" id="cb3-11" title="11">    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_len</span>(<span class="kw">nrow</span>(df)))</a>
<a class="sourceLine" id="cb3-12" title="12">    {</a>
<a class="sourceLine" id="cb3-13" title="13">        infile.fasta &lt;-<span class="st"> </span>df[i,F]</a>
<a class="sourceLine" id="cb3-14" title="14">        <span class="kw">cat</span>(<span class="st">&#39;</span><span class="ch">\n</span><span class="st">process &#39;</span>,infile.fasta)</a>
<a class="sourceLine" id="cb3-15" title="15">        sq                  &lt;-<span class="st"> </span><span class="kw">read.dna</span>(infile.fasta, <span class="dt">format=</span><span class="st">&#39;fa&#39;</span>)  </a>
<a class="sourceLine" id="cb3-16" title="16">        <span class="kw">set.seed</span>(seed)</a>
<a class="sourceLine" id="cb3-17" title="17">        <span class="cf">for</span>(b <span class="cf">in</span> <span class="dv">0</span><span class="op">:</span>bsn)</a>
<a class="sourceLine" id="cb3-18" title="18">        {</a>
<a class="sourceLine" id="cb3-19" title="19">            outfile &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">.fasta&#39;</span>,<span class="kw">paste0</span>(<span class="st">&#39;_&#39;</span>,<span class="kw">sprintf</span>(<span class="st">&quot;%03d&quot;</span>,b),<span class="st">&#39;.fasta&#39;</span>),infile.fasta)</a>
<a class="sourceLine" id="cb3-20" title="20">            outfile &lt;-<span class="st"> </span><span class="kw">file.path</span>(outdir,<span class="kw">basename</span>(outfile))</a>
<a class="sourceLine" id="cb3-21" title="21">            bs &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(sq)</a>
<a class="sourceLine" id="cb3-22" title="22">            <span class="cf">if</span>(b<span class="op">&gt;</span><span class="dv">0</span>)</a>
<a class="sourceLine" id="cb3-23" title="23">            {</a>
<a class="sourceLine" id="cb3-24" title="24">                bs&lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(sq), <span class="kw">ncol</span>(sq), <span class="dt">replace=</span><span class="ot">TRUE</span>) </a>
<a class="sourceLine" id="cb3-25" title="25">            }           </a>
<a class="sourceLine" id="cb3-26" title="26">            <span class="kw">write.dna</span>(sq[,bs], <span class="dt">file=</span>outfile, <span class="dt">format=</span><span class="st">&#39;fa&#39;</span>, <span class="dt">colsep=</span><span class="st">&#39;&#39;</span>, <span class="dt">nbcol=</span><span class="op">-</span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb3-27" title="27">        }       </a>
<a class="sourceLine" id="cb3-28" title="28">    }   </a>
<a class="sourceLine" id="cb3-29" title="29">}</a></code></pre></div>
</div>
<div id="step-3-make-phylogenetic-trees" class="section level2">
<h2><strong>Step 3 (make phylogenetic trees)</strong></h2>
<p>Next, I reconstruct viral phylogenies, one for each bootstrap alignment. I have used Fasttree, primarily because I just want to find fairly large and distinct groups of viral lineages. I accept that some of the details, and in particular the branch lengths, may not be quite right. We have about 50,000 subtype B sequences, and for those Fasttree takes a couple of hours to return a viral phylogeny. For the other subtypes, the phylogenies are pretty much reconstructed instantly, and we could use RAxML without problems.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1">seattle.<span class="fl">191017.</span>fastree&lt;-<span class="st"> </span><span class="cf">function</span>()</a>
<a class="sourceLine" id="cb4-2" title="2">{   </a>
<a class="sourceLine" id="cb4-3" title="3">    <span class="kw">require</span>(big.phylo)</a>
<a class="sourceLine" id="cb4-4" title="4">    <span class="kw">require</span>(data.table)</a>
<a class="sourceLine" id="cb4-5" title="5">    </a>
<a class="sourceLine" id="cb4-6" title="6">    home &lt;-<span class="st"> &#39;~/Box Sync/OR_Work/Seattle/analysis_191017&#39;</span></a>
<a class="sourceLine" id="cb4-7" title="7">    home &lt;-<span class="st"> &#39;/rds/general/project/ratmann_seattle_data_analysis/live/olli_191017&#39;</span></a>
<a class="sourceLine" id="cb4-8" title="8">    </a>
<a class="sourceLine" id="cb4-9" title="9">    indir &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;alignments_bs&#39;</span>)</a>
<a class="sourceLine" id="cb4-10" title="10">    outdir &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;fasttree&#39;</span>)</a>
<a class="sourceLine" id="cb4-11" title="11">    </a>
<a class="sourceLine" id="cb4-12" title="12">    <span class="co">#   get UNIX commands for all trees to be processed</span></a>
<a class="sourceLine" id="cb4-13" title="13">    df &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">FIN=</span><span class="kw">list.files</span>(indir, <span class="dt">pattern=</span><span class="st">&#39;fasta$&#39;</span>,<span class="dt">full.names=</span><span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb4-14" title="14">    df &lt;-<span class="st"> </span><span class="kw">subset</span>(df, <span class="kw">grepl</span>(<span class="st">&#39;ndrm&#39;</span>,FIN))</a>
<a class="sourceLine" id="cb4-15" title="15">    df[, ST<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;.*_Subtype([A-Z0-9]+)_.*&#39;</span>,<span class="st">&#39;</span><span class="ch">\\</span><span class="st">1&#39;</span>,<span class="kw">basename</span>(FIN))]</a>
<a class="sourceLine" id="cb4-16" title="16">    df[, FOUT<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">file.path</span>(outdir,<span class="kw">gsub</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">.fasta&#39;</span>,<span class="st">&#39;_ft.newick&#39;</span>,<span class="kw">basename</span>(FIN)))]</a>
<a class="sourceLine" id="cb4-17" title="17">    cmds &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&#39;list&#39;</span>, <span class="kw">nrow</span>(df))</a>
<a class="sourceLine" id="cb4-18" title="18">    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_len</span>(<span class="kw">nrow</span>(df)))</a>
<a class="sourceLine" id="cb4-19" title="19">    {</a>
<a class="sourceLine" id="cb4-20" title="20">        infile.fasta &lt;-<span class="st"> </span>df[i,FIN]</a>
<a class="sourceLine" id="cb4-21" title="21">        outfile &lt;-<span class="st"> </span>df[i,FOUT]           </a>
<a class="sourceLine" id="cb4-22" title="22">        tmp &lt;-<span class="st"> </span><span class="kw">cmd.fasttree</span>(infile.fasta, <span class="dt">outfile=</span>outfile, <span class="dt">pr.args=</span><span class="st">&#39;-nt -gtr -gamma&#39;</span>, <span class="dt">check.binary=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-23" title="23">        cmds[[i]] &lt;-<span class="st"> </span>tmp</a>
<a class="sourceLine" id="cb4-24" title="24">    }</a>
<a class="sourceLine" id="cb4-25" title="25">    df[, CMD<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">unlist</span>(cmds)]    </a>
<a class="sourceLine" id="cb4-26" title="26">    </a>
<a class="sourceLine" id="cb4-27" title="27">    <span class="co">#   submit jobs like this one:</span></a>
<a class="sourceLine" id="cb4-28" title="28">    <span class="kw">cat</span>(df[<span class="dv">1</span>,CMD])</a>
<a class="sourceLine" id="cb4-29" title="29">    </a>
<a class="sourceLine" id="cb4-30" title="30">    <span class="co">#   run on HPC as array job</span></a>
<a class="sourceLine" id="cb4-31" title="31">    df &lt;-<span class="st"> </span><span class="kw">subset</span>(df, ST<span class="op">==</span><span class="st">&#39;B&#39;</span>)</a>
<a class="sourceLine" id="cb4-32" title="32">    df[, CASE_ID<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(df)]</a>
<a class="sourceLine" id="cb4-33" title="33">    </a>
<a class="sourceLine" id="cb4-34" title="34">    <span class="co">#   make PBS header</span></a>
<a class="sourceLine" id="cb4-35" title="35">    hpc.load    &lt;-<span class="st"> &quot;module load R/3.3.3&quot;</span></a>
<a class="sourceLine" id="cb4-36" title="36">    hpc.select  &lt;-<span class="st"> </span><span class="dv">1</span>                        <span class="co"># number of nodes</span></a>
<a class="sourceLine" id="cb4-37" title="37">    hpc.nproc   &lt;-<span class="st"> </span><span class="dv">1</span>                        <span class="co"># number of processors on node</span></a>
<a class="sourceLine" id="cb4-38" title="38">    hpc.walltime&lt;-<span class="st"> </span><span class="dv">923</span>                      <span class="co"># walltime</span></a>
<a class="sourceLine" id="cb4-39" title="39">    hpc.q       &lt;-<span class="st"> &quot;pqeelab&quot;</span>                        <span class="co"># PBS queue</span></a>
<a class="sourceLine" id="cb4-40" title="40">    hpc.mem     &lt;-<span class="st"> &quot;6gb&quot;</span>                    <span class="co"># RAM</span></a>
<a class="sourceLine" id="cb4-41" title="41">    hpc.array   &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(df<span class="op">$</span>CASE_ID))   <span class="co"># number of runs for job array</span></a>
<a class="sourceLine" id="cb4-42" title="42">    pbshead     &lt;-<span class="st"> &quot;#!/bin/sh&quot;</span></a>
<a class="sourceLine" id="cb4-43" title="43">    tmp         &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;#PBS -l walltime=&quot;</span>, hpc.walltime, <span class="st">&quot;:59:00,pcput=&quot;</span>, hpc.walltime, <span class="st">&quot;:45:00&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb4-44" title="44">    pbshead     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb4-45" title="45">    tmp         &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;#PBS -l select=&quot;</span>, hpc.select, <span class="st">&quot;:ncpus=&quot;</span>, hpc.nproc,<span class="st">&quot;:mem=&quot;</span>, hpc.mem, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb4-46" title="46">    pbshead     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb4-47" title="47">    pbshead     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="st">&quot;#PBS -j oe&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>) </a>
<a class="sourceLine" id="cb4-48" title="48">    <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.na</span>(hpc.array))</a>
<a class="sourceLine" id="cb4-49" title="49">        pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">#PBS -J 1-&quot;</span>, hpc.array, <span class="dt">sep=</span><span class="st">&#39;&#39;</span>)        </a>
<a class="sourceLine" id="cb4-50" title="50">    <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.na</span>(hpc.q)) </a>
<a class="sourceLine" id="cb4-51" title="51">        pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="kw">paste</span>(<span class="st">&quot;#PBS -q&quot;</span>, hpc.q), <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb4-52" title="52">    pbshead     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, hpc.load, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)         </a>
<a class="sourceLine" id="cb4-53" title="53">    <span class="co">#   make array job</span></a>
<a class="sourceLine" id="cb4-54" title="54">    cmd     &lt;-<span class="st"> </span>df[, <span class="kw">list</span>(<span class="dt">CASE=</span><span class="kw">paste0</span>(CASE_ID,<span class="st">&#39;)</span><span class="ch">\n</span><span class="st">&#39;</span>,CMD,<span class="st">&#39;;;</span><span class="ch">\n</span><span class="st">&#39;</span>)), by=<span class="st">&#39;CASE_ID&#39;</span>]</a>
<a class="sourceLine" id="cb4-55" title="55">    cmd     &lt;-<span class="st"> </span>cmd[, <span class="kw">paste0</span>(<span class="st">&#39;case $PBS_ARRAY_INDEX in</span><span class="ch">\n</span><span class="st">&#39;</span>,<span class="kw">paste0</span>(CASE, <span class="dt">collapse=</span><span class="st">&#39;&#39;</span>),<span class="st">&#39;esac&#39;</span>)]         </a>
<a class="sourceLine" id="cb4-56" title="56">    cmd     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead,cmd,<span class="dt">sep=</span><span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>)  </a>
<a class="sourceLine" id="cb4-57" title="57">    <span class="co">#   submit job</span></a>
<a class="sourceLine" id="cb4-58" title="58">    outfile     &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;:&#39;</span>,<span class="st">&#39;&#39;</span>,<span class="kw">paste</span>(<span class="st">&quot;trs&quot;</span>,<span class="kw">paste</span>(<span class="kw">strsplit</span>(<span class="kw">date</span>(),<span class="dt">split=</span><span class="st">&#39; &#39;</span>)[[<span class="dv">1</span>]],<span class="dt">collapse=</span><span class="st">&#39;_&#39;</span>,<span class="dt">sep=</span><span class="st">&#39;&#39;</span>),<span class="st">&#39;sh&#39;</span>,<span class="dt">sep=</span><span class="st">&#39;.&#39;</span>))</a>
<a class="sourceLine" id="cb4-59" title="59">    outfile     &lt;-<span class="st"> </span><span class="kw">file.path</span>(outdir, outfile)</a>
<a class="sourceLine" id="cb4-60" title="60">    <span class="kw">cat</span>(cmd, <span class="dt">file=</span>outfile)</a>
<a class="sourceLine" id="cb4-61" title="61">    cmd         &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;qsub&quot;</span>, outfile)</a>
<a class="sourceLine" id="cb4-62" title="62">    <span class="kw">cat</span>(cmd)</a>
<a class="sourceLine" id="cb4-63" title="63">    <span class="kw">cat</span>(<span class="kw">system</span>(cmd, <span class="dt">intern=</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb4-64" title="64">}</a></code></pre></div>
</div>
<div id="step-4-make-ancestral-state-reconstruction" class="section level2">
<h2><strong>Step 4 (make ancestral state reconstruction)</strong></h2>
<p>Next, I want to associate viral lineages to geographic locations, and then isolate distinct sub-trees that are geographically associated to King County/Seattle. We know the sampling location of all viral sequences, which is the only information needed for reconstructing phylogeographic ancestral states. Every Seattle sequence will be in one sub-tree, so all the sequence data will be kept in the analysis, in contrast to phylogenetic clustering approaches.</p>
<p>There are several phylogeographic models that we could use for phylogeographic reconstruction. Parsimony-based models are fast and directly applicable to the large subtype-B tree. Another option could be Lemey’s discrete-state diffusion model, but we would have to cut large phylogenies into smaller pieces to make it computationally feasible. The parsimony-based models are more sensitive to sequence sampling bias so we need to include as many sequences from LANL as possible to get the parsimony-based reconstruction roughly right.</p>
<p>There is one more twist to the analyses. From an epidemiologic perspective, I always consider spread among MSM and heterosexual individuals (HSX) as separate things. This can be easily accommodated in our phylogeographic analysis, because the tip states can be anything, they do not have to be sampling locations. So let us define the tip states in terms of sampling location and transmission group. Specifically, for the Seattle sequences, we consider KC-HSX and KC-non-HSX. For the LANL sequences, we consider larger world regions such as South America, Europe, sub-Saharan Africa, etc. Then, we repeat analysis using KC-MSM and KC-non-MSM.</p>
<p>First, I prepared the sampling locations for each viral sequence:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">seattle.<span class="fl">191017.</span>sequence.labels&lt;-<span class="st"> </span><span class="cf">function</span>()</a>
<a class="sourceLine" id="cb5-2" title="2">{</a>
<a class="sourceLine" id="cb5-3" title="3">    <span class="kw">require</span>(data.table)</a>
<a class="sourceLine" id="cb5-4" title="4">    <span class="kw">require</span>(ape)</a>
<a class="sourceLine" id="cb5-5" title="5">    <span class="kw">require</span>(tidyverse)</a>
<a class="sourceLine" id="cb5-6" title="6">    </a>
<a class="sourceLine" id="cb5-7" title="7">    home &lt;-<span class="st"> &#39;~/Box Sync/OR_Work/Seattle&#39;</span></a>
<a class="sourceLine" id="cb5-8" title="8">    infile.indinfo &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;PHSKC-2018-07-09&#39;</span>,<span class="st">&#39;person.rds&#39;</span>)</a>
<a class="sourceLine" id="cb5-9" title="9">    infile.seqinfo &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;PHSKC-2018-07-09&#39;</span>,<span class="st">&#39;sequences_meta.rds&#39;</span>)   </a>
<a class="sourceLine" id="cb5-10" title="10">    infile.countryinfo &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;analysis_191017&#39;</span>,<span class="st">&#39;misc/Country_db.rds&#39;</span>)</a>
<a class="sourceLine" id="cb5-11" title="11">    infiles.lanl &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;analysis_191017&#39;</span>,<span class="st">&#39;alignments&#39;</span>,<span class="kw">c</span>(<span class="st">&#39;180709_LANL_Subtype01AE_mafft.fasta&#39;</span>,<span class="st">&#39;180709_LANL_SubtypeA1_mafft.fasta&#39;</span>,<span class="st">&#39;180709_LANL_SubtypeB_mafft.fasta&#39;</span>,<span class="st">&#39;180709_LANL_SubtypeC_mafft.fasta&#39;</span>))</a>
<a class="sourceLine" id="cb5-12" title="12">    infile.subtype &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;analysis_191017&#39;</span>,<span class="st">&#39;misc&#39;</span>,<span class="st">&#39;180709_Subtype.csv&#39;</span>)</a>
<a class="sourceLine" id="cb5-13" title="13">    outfile.base &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;analysis_191017&#39;</span>,<span class="st">&#39;misc&#39;</span>,<span class="st">&#39;180709_&#39;</span>)</a>
<a class="sourceLine" id="cb5-14" title="14">    </a>
<a class="sourceLine" id="cb5-15" title="15">    <span class="co">#</span></a>
<a class="sourceLine" id="cb5-16" title="16">    <span class="co"># collect country codes</span></a>
<a class="sourceLine" id="cb5-17" title="17">    <span class="co">#</span></a>
<a class="sourceLine" id="cb5-18" title="18">    dco &lt;-<span class="st"> </span><span class="kw">readRDS</span>(infile.countryinfo)</a>
<a class="sourceLine" id="cb5-19" title="19">    dco &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">seq_len</span>(<span class="kw">length</span>(dco)), <span class="cf">function</span>(i) <span class="kw">tibble</span>(<span class="dt">WRLD=</span><span class="kw">names</span>(dco)[i], <span class="dt">CNTRY=</span>dco[[i]]))</a>
<a class="sourceLine" id="cb5-20" title="20">    dco &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">&#39;rbind&#39;</span>,dco)</a>
<a class="sourceLine" id="cb5-21" title="21">    dco &lt;-<span class="st"> </span>dco <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">WRLD=</span><span class="kw">gsub</span>(<span class="st">&#39;Unkno&#39;</span>,<span class="st">&#39;Unknown&#39;</span>,<span class="kw">gsub</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">.|</span><span class="ch">\\</span><span class="st">_&#39;</span>,<span class="st">&#39;&#39;</span>,WRLD))) </a>
<a class="sourceLine" id="cb5-22" title="22">    tmp &lt;-<span class="st"> </span><span class="kw">tibble</span>(  <span class="dt">CNTRY=</span> <span class="kw">c</span>(<span class="st">&quot;BO&quot;</span>, <span class="st">&quot;SD&quot;</span>, <span class="st">&quot;EC&quot;</span>, <span class="st">&quot;AG&quot;</span>, <span class="st">&quot;DM&quot;</span>, <span class="st">&quot;GD&quot;</span>, <span class="st">&quot;GY&quot;</span>, <span class="st">&quot;VC&quot;</span>, <span class="st">&quot;SR&quot;</span>, <span class="st">&quot;LC&quot;</span>, <span class="st">&quot;UZ&quot;</span>, <span class="st">&quot;AZ&quot;</span>, <span class="st">&quot;GH&quot;</span>, <span class="st">&quot;HT&quot;</span>, <span class="st">&quot;LV&quot;</span>, <span class="st">&quot;BZ&quot;</span>, <span class="st">&quot;MN&quot;</span>, <span class="st">&quot;GF&quot;</span>, <span class="st">&quot;TJ&quot;</span>),</a>
<a class="sourceLine" id="cb5-23" title="23">                    <span class="dt">NAME=</span> <span class="kw">c</span>(<span class="st">&quot;Bolivia&quot;</span>,<span class="st">&quot;Sudan&quot;</span>,<span class="st">&quot;Ecuador&quot;</span>,<span class="st">&quot;Antigua&quot;</span>,<span class="st">&quot;Dominica&quot;</span>,<span class="st">&quot;Grenada&quot;</span>,<span class="st">&quot;Guyana&quot;</span>,<span class="st">&quot;Saint Vincent&quot;</span>,<span class="st">&quot;Suriname&quot;</span>,<span class="st">&quot;Saint Lucia&quot;</span>,<span class="st">&quot;Uzbekistan&quot;</span>,<span class="st">&quot;Azerbaijan&quot;</span>,<span class="st">&quot;Ghana&quot;</span>,<span class="st">&quot;Haiti&quot;</span>,<span class="st">&quot;Latvia&quot;</span>,<span class="st">&quot;Belize&quot;</span>,<span class="st">&quot;Mongolia&quot;</span>,<span class="st">&quot;French Guiana&quot;</span>,<span class="st">&quot;Tajikistan&quot;</span>),</a>
<a class="sourceLine" id="cb5-24" title="24">                    <span class="dt">WRLD=</span> <span class="kw">c</span>(<span class="st">&quot;SMAm&quot;</span>,<span class="st">&quot;NorAf&quot;</span>,<span class="st">&quot;SMAm&quot;</span>,<span class="st">&quot;SMAm&quot;</span>,<span class="st">&quot;SMAm&quot;</span>,<span class="st">&quot;SMAm&quot;</span>,<span class="st">&quot;SMAm&quot;</span>,<span class="st">&quot;SMAm&quot;</span>,<span class="st">&quot;SMAm&quot;</span>,<span class="st">&quot;SMAm&quot;</span>,<span class="st">&quot;Asia&quot;</span>,<span class="st">&quot;Asia&quot;</span>,<span class="st">&quot;SSA&quot;</span>,<span class="st">&quot;SMAm&quot;</span>,<span class="st">&quot;Europe&quot;</span>,<span class="st">&quot;SMAm&quot;</span>,<span class="st">&quot;Asia&quot;</span>,<span class="st">&quot;SMAm&quot;</span>,<span class="st">&quot;Asia&quot;</span>)</a>
<a class="sourceLine" id="cb5-25" title="25">                    )</a>
<a class="sourceLine" id="cb5-26" title="26">    dco &lt;-<span class="st"> </span>tmp <span class="op">%&gt;%</span><span class="st">  </span></a>
<a class="sourceLine" id="cb5-27" title="27"><span class="st">        </span><span class="kw">select</span>(WRLD, CNTRY) <span class="op">%&gt;%</span><span class="st">         </span></a>
<a class="sourceLine" id="cb5-28" title="28"><span class="st">        </span><span class="kw">rbind</span>(dco) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-29" title="29"><span class="st">        </span><span class="kw">mutate</span>(<span class="dt">WRLD:=</span> <span class="kw">case_when</span>(CNTRY<span class="op">==</span><span class="st">&#39;MX&#39;</span><span class="op">~</span><span class="st">&quot;SMAm&quot;</span>,CNTRY<span class="op">==</span><span class="st">&#39;CA&#39;</span><span class="op">~</span><span class="st">&#39;Canada&#39;</span>,CNTRY<span class="op">==</span><span class="st">&#39;GL&#39;</span><span class="op">~</span><span class="st">&#39;Canada&#39;</span>,CNTRY<span class="op">!=</span><span class="st">&#39;GL&#39;</span><span class="op">&amp;</span>CNTRY<span class="op">!=</span><span class="st">&#39;CA&#39;</span><span class="op">&amp;</span>CNTRY<span class="op">!=</span><span class="st">&#39;MX&#39;</span><span class="op">~</span>WRLD)) <span class="op">%&gt;%</span><span class="st">     </span></a>
<a class="sourceLine" id="cb5-30" title="30"><span class="st">        </span><span class="kw">arrange</span>(WRLD, CNTRY)</a>
<a class="sourceLine" id="cb5-31" title="31">    <span class="kw">table</span>(dco<span class="op">$</span>WRLD, dco<span class="op">$</span>CNTRY)  </a>
<a class="sourceLine" id="cb5-32" title="32"></a>
<a class="sourceLine" id="cb5-33" title="33">    <span class="co">#</span></a>
<a class="sourceLine" id="cb5-34" title="34">    <span class="co"># read Seattle indidivual data </span></a>
<a class="sourceLine" id="cb5-35" title="35">    <span class="co">#   </span></a>
<a class="sourceLine" id="cb5-36" title="36">    dind &lt;-<span class="st"> </span><span class="kw">readRDS</span>(infile.indinfo) </a>
<a class="sourceLine" id="cb5-37" title="37">    dind &lt;-<span class="st"> </span>dind <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-38" title="38"><span class="st">            </span><span class="kw">select</span>(newnum, race, b_yr, birthCountry, gender, sex, sex_with_male_and_female, transm, rsh_county_name, rsa_county_name, cur_county_name) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-39" title="39"><span class="st">            </span><span class="kw">mutate</span>( <span class="dt">birthCountry2:=</span> <span class="kw">gsub</span>(<span class="st">&#39;^</span><span class="ch">\\</span><span class="st">(([A-Z0-9]+)</span><span class="ch">\\</span><span class="st">).*$&#39;</span>,<span class="st">&#39;</span><span class="ch">\\</span><span class="st">1&#39;</span>, birthCountry),</a>
<a class="sourceLine" id="cb5-40" title="40">                    <span class="dt">county:=</span> <span class="kw">case_when</span>( rsh_county_name<span class="op">!=</span><span class="st">&quot;KING CO.&quot;</span><span class="op">&amp;</span>rsa_county_name<span class="op">!=</span><span class="st">&quot;KING CO.&quot;</span><span class="op">&amp;</span>cur_county_name<span class="op">!=</span><span class="st">&quot;KING CO.&quot;</span><span class="op">~</span><span class="st">&#39;other&#39;</span>,</a>
<a class="sourceLine" id="cb5-41" title="41">                                        rsh_county_name<span class="op">==</span><span class="st">&quot;KING CO.&quot;</span><span class="op">|</span>rsa_county_name<span class="op">==</span><span class="st">&quot;KING CO.&quot;</span><span class="op">|</span>cur_county_name<span class="op">==</span><span class="st">&quot;KING CO.&quot;</span><span class="op">~</span><span class="st">&#39;king&#39;</span>),</a>
<a class="sourceLine" id="cb5-42" title="42">                    <span class="dt">Gender2:=</span> <span class="kw">case_when</span>(gender<span class="op">==</span><span class="st">&#39;FM&#39;</span><span class="op">~</span><span class="st">&#39;Trns&#39;</span>,</a>
<a class="sourceLine" id="cb5-43" title="43">                                        gender<span class="op">==</span><span class="st">&#39;MF&#39;</span><span class="op">~</span><span class="st">&#39;Trns&#39;</span>,</a>
<a class="sourceLine" id="cb5-44" title="44">                                        gender<span class="op">==</span><span class="st">&#39;F&#39;</span><span class="op">&amp;</span>sex<span class="op">==</span><span class="st">&#39;F&#39;</span><span class="op">~</span><span class="st">&#39;F&#39;</span>,</a>
<a class="sourceLine" id="cb5-45" title="45">                                        gender<span class="op">==</span><span class="st">&#39;F&#39;</span><span class="op">&amp;</span>sex<span class="op">==</span><span class="st">&#39;M&#39;</span><span class="op">~</span><span class="st">&#39;Trns&#39;</span>,</a>
<a class="sourceLine" id="cb5-46" title="46">                                        gender<span class="op">==</span><span class="st">&#39;M&#39;</span><span class="op">&amp;</span>sex<span class="op">==</span><span class="st">&#39;M&#39;</span><span class="op">~</span><span class="st">&#39;M&#39;</span>,</a>
<a class="sourceLine" id="cb5-47" title="47">                                        gender<span class="op">==</span><span class="st">&#39;M&#39;</span><span class="op">&amp;</span>sex<span class="op">==</span><span class="st">&#39;F&#39;</span><span class="op">~</span><span class="st">&#39;Trns&#39;</span>),</a>
<a class="sourceLine" id="cb5-48" title="48">                    <span class="dt">transm2:=</span> <span class="kw">case_when</span>(<span class="kw">grepl</span>(<span class="st">&#39;MSM&#39;</span>,transm)<span class="op">~</span><span class="st">&#39;MSM&#39;</span>,</a>
<a class="sourceLine" id="cb5-49" title="49">                                        <span class="kw">grepl</span>(<span class="st">&#39;HETERO&#39;</span>,transm)<span class="op">~</span><span class="st">&#39;HSX&#39;</span>,</a>
<a class="sourceLine" id="cb5-50" title="50">                                        <span class="kw">grepl</span>(<span class="st">&#39;BLOOD|PERINAT|OTHER|IDU&#39;</span>,transm)<span class="op">~</span><span class="st">&#39;OTH&#39;</span>,</a>
<a class="sourceLine" id="cb5-51" title="51">                                        <span class="kw">grepl</span>(<span class="st">&#39;UNKNOWN&#39;</span>,transm)<span class="op">~</span><span class="st">&#39;UNKNOWN&#39;</span>),</a>
<a class="sourceLine" id="cb5-52" title="52">                    <span class="dt">race2:=</span> <span class="kw">case_when</span>( <span class="kw">grepl</span>(<span class="st">&#39;Black&#39;</span>,race)<span class="op">~</span><span class="st">&#39;Black&#39;</span>,</a>
<a class="sourceLine" id="cb5-53" title="53">                                    <span class="kw">grepl</span>(<span class="st">&#39;White&#39;</span>,race)<span class="op">~</span><span class="st">&#39;White&#39;</span>,</a>
<a class="sourceLine" id="cb5-54" title="54">                                    <span class="kw">grepl</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">(1</span><span class="ch">\\</span><span class="st">)Hispanic&#39;</span>,race)<span class="op">~</span><span class="st">&#39;Hispanic&#39;</span>,</a>
<a class="sourceLine" id="cb5-55" title="55">                                    <span class="kw">grepl</span>(<span class="st">&#39;Hawaiian|Indian|Asian|Multi-race&#39;</span>,race)<span class="op">~</span><span class="st">&#39;Other&#39;</span>,</a>
<a class="sourceLine" id="cb5-56" title="56">                                    <span class="kw">grepl</span>(<span class="st">&#39;Unknown&#39;</span>,race)<span class="op">~</span><span class="st">&#39;Unknown&#39;</span></a>
<a class="sourceLine" id="cb5-57" title="57">                                    )) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-58" title="58"><span class="st">            </span><span class="kw">mutate</span>( <span class="dt">birthCountry2:=</span> <span class="kw">case_when</span>(  birthCountry2<span class="op">==</span><span class="st">&#39;&#39;</span><span class="op">~</span><span class="st">&#39;Unknown&#39;</span>,</a>
<a class="sourceLine" id="cb5-59" title="59">                                                birthCountry2<span class="op">==</span><span class="st">&#39;X98&#39;</span><span class="op">~</span><span class="st">&#39;Unknown&#39;</span>,</a>
<a class="sourceLine" id="cb5-60" title="60">                                                birthCountry2<span class="op">==</span><span class="st">&#39;X99&#39;</span><span class="op">~</span><span class="st">&#39;Unknown&#39;</span>,</a>
<a class="sourceLine" id="cb5-61" title="61">                                                birthCountry2<span class="op">!=</span><span class="st">&#39;&#39;</span><span class="op">&amp;</span>birthCountry2<span class="op">!=</span><span class="st">&#39;X98&#39;</span><span class="op">&amp;</span>birthCountry2<span class="op">!=</span><span class="st">&#39;X99&#39;</span><span class="op">~</span>birthCountry2</a>
<a class="sourceLine" id="cb5-62" title="62">                                                ))</a>
<a class="sourceLine" id="cb5-63" title="63">    <span class="co">#   setdiff( dind$birthCountry2, dco$CNTRY )    </span></a>
<a class="sourceLine" id="cb5-64" title="64">    <span class="co">#   let s not resolve country of birth to world region for now..</span></a>
<a class="sourceLine" id="cb5-65" title="65">    <span class="cf">for</span>(x <span class="cf">in</span> <span class="kw">colnames</span>(dind))</a>
<a class="sourceLine" id="cb5-66" title="66">     <span class="kw">attr</span>(dind[[x]], <span class="st">&quot;label&quot;</span>) &lt;-<span class="st"> </span><span class="ot">NULL</span>           </a>
<a class="sourceLine" id="cb5-67" title="67">    <span class="co">#           </span></a>
<a class="sourceLine" id="cb5-68" title="68">    <span class="co">#table(dind$Gender2, dind$transm2, dind$sex_with_male_and_female) # sex_with_male_and_female does not look very reliable</span></a>
<a class="sourceLine" id="cb5-69" title="69"></a>
<a class="sourceLine" id="cb5-70" title="70">    <span class="co">#</span></a>
<a class="sourceLine" id="cb5-71" title="71">    <span class="co"># read Seattle sequence data </span></a>
<a class="sourceLine" id="cb5-72" title="72">    <span class="co">#   </span></a>
<a class="sourceLine" id="cb5-73" title="73">    dseq &lt;-<span class="st"> </span><span class="kw">readRDS</span>(infile.seqinfo)</a>
<a class="sourceLine" id="cb5-74" title="74">    dseq &lt;-<span class="st"> </span>dseq <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(type<span class="op">==</span><span class="st">&#39;PR/RT&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(seqID, newnum, seqy)</a>
<a class="sourceLine" id="cb5-75" title="75">    <span class="cf">for</span>(x <span class="cf">in</span> <span class="kw">colnames</span>(dseq))</a>
<a class="sourceLine" id="cb5-76" title="76">        <span class="kw">attr</span>(dseq[[x]], <span class="st">&quot;label&quot;</span>) &lt;-<span class="st"> </span><span class="ot">NULL</span>    </a>
<a class="sourceLine" id="cb5-77" title="77">    dseq &lt;-<span class="st"> </span>dind <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">inner_join</span>(dseq, <span class="dt">by=</span><span class="st">&#39;newnum&#39;</span>)  </a>
<a class="sourceLine" id="cb5-78" title="78">    dseq &lt;-<span class="st"> </span>dseq <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(newnum, seqID, seqy, county, Gender2, transm2, race2, birthCountry2)    </a>
<a class="sourceLine" id="cb5-79" title="79">    tmp &lt;-<span class="st"> </span><span class="kw">as_tibble</span>(<span class="kw">read.csv</span>(infile.subtype, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-80" title="80"><span class="st">            </span><span class="kw">rename</span>(<span class="dt">seqID=</span>name, <span class="dt">ST=</span>subtype) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-81" title="81"><span class="st">            </span><span class="kw">select</span>(seqID, ST) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-82" title="82"><span class="st">            </span><span class="kw">mutate</span>(<span class="dt">seqID=</span> <span class="kw">gsub</span>(<span class="st">&#39;PRRT&#39;</span>,<span class="st">&#39;PR/RT&#39;</span>,seqID))</a>
<a class="sourceLine" id="cb5-83" title="83">    dseq &lt;-<span class="st"> </span>dseq <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">left_join</span>(tmp, <span class="dt">by=</span><span class="st">&#39;seqID&#39;</span>)</a>
<a class="sourceLine" id="cb5-84" title="84">    </a>
<a class="sourceLine" id="cb5-85" title="85">    <span class="co">#</span></a>
<a class="sourceLine" id="cb5-86" title="86">    <span class="co">#   collect LANL labels</span></a>
<a class="sourceLine" id="cb5-87" title="87">    <span class="co">#</span></a>
<a class="sourceLine" id="cb5-88" title="88">    dl &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">seq_along</span>(infiles.lanl), <span class="cf">function</span>(i) <span class="kw">rownames</span>(<span class="kw">read.dna</span>(infiles.lanl[i],<span class="dt">format=</span><span class="st">&#39;fa&#39;</span>)))</a>
<a class="sourceLine" id="cb5-89" title="89">    dl &lt;-<span class="st"> </span><span class="kw">tibble</span>(<span class="dt">TAXA=</span><span class="kw">unlist</span>(dl)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-90" title="90"><span class="st">            </span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">grepl</span>(<span class="st">&#39;PR</span><span class="ch">\\</span><span class="st">/RT&#39;</span>,TAXA) <span class="op">&amp;</span><span class="st"> </span>TAXA<span class="op">!=</span><span class="st">&#39;HXB2&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-91" title="91"><span class="st">            </span><span class="kw">mutate</span>( <span class="dt">ST:=</span> <span class="kw">sapply</span>(<span class="kw">strsplit</span>(TAXA,<span class="st">&#39;.&#39;</span>,<span class="dt">fixed=</span><span class="ot">TRUE</span>),<span class="st">&#39;[[&#39;</span>,<span class="dv">1</span>),</a>
<a class="sourceLine" id="cb5-92" title="92">                    <span class="dt">CNTRY:=</span> <span class="kw">sapply</span>(<span class="kw">strsplit</span>(TAXA,<span class="st">&#39;.&#39;</span>,<span class="dt">fixed=</span><span class="ot">TRUE</span>),<span class="st">&#39;[[&#39;</span>,<span class="dv">2</span>),</a>
<a class="sourceLine" id="cb5-93" title="93">                    <span class="dt">YEAR:=</span> <span class="kw">sapply</span>(<span class="kw">strsplit</span>(TAXA,<span class="st">&#39;.&#39;</span>,<span class="dt">fixed=</span><span class="ot">TRUE</span>),<span class="st">&#39;[[&#39;</span>,<span class="dv">3</span>),</a>
<a class="sourceLine" id="cb5-94" title="94">                    <span class="dt">GENBANK:=</span> <span class="kw">sapply</span>(<span class="kw">strsplit</span>(TAXA,<span class="st">&#39;.&#39;</span>,<span class="dt">fixed=</span><span class="ot">TRUE</span>),<span class="st">&#39;[[&#39;</span>,<span class="dv">5</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-95" title="95"><span class="st">            </span><span class="kw">distinct</span>()</a>
<a class="sourceLine" id="cb5-96" title="96">    dl &lt;-<span class="st"> </span>dl <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-97" title="97"><span class="st">            </span><span class="kw">filter</span>(<span class="kw">grepl</span>(<span class="st">&#39;HXB2&#39;</span>,TAXA)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-98" title="98"><span class="st">            </span><span class="kw">mutate</span>(<span class="dt">TAXA:=</span><span class="st">&#39;HXB2&#39;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-99" title="99"><span class="st">            </span><span class="kw">rbind</span>(dl)   </a>
<a class="sourceLine" id="cb5-100" title="100">    <span class="co">#</span></a>
<a class="sourceLine" id="cb5-101" title="101">    outfile &lt;-<span class="st"> </span><span class="kw">paste0</span>(outfile.base,<span class="st">&#39;sequence_labels.rda&#39;</span>)</a>
<a class="sourceLine" id="cb5-102" title="102">    <span class="kw">save</span>(dseq, dind, dl, dco, <span class="dt">file=</span> outfile)</a>
<a class="sourceLine" id="cb5-103" title="103">    </a>
<a class="sourceLine" id="cb5-104" title="104">    <span class="co">#</span></a>
<a class="sourceLine" id="cb5-105" title="105">    <span class="co">#   make sequence labels for KCMSM</span></a>
<a class="sourceLine" id="cb5-106" title="106">    <span class="co">#</span></a>
<a class="sourceLine" id="cb5-107" title="107">    tmp &lt;-<span class="st"> </span>dseq <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rename</span>(<span class="dt">TAXA:=</span> seqID) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-108" title="108"><span class="st">            </span><span class="kw">mutate</span>(<span class="dt">WRLD:=</span> <span class="kw">case_when</span>(county<span class="op">==</span><span class="st">&#39;king&#39;</span><span class="op">&amp;</span>transm2<span class="op">==</span><span class="st">&#39;MSM&#39;</span><span class="op">~</span><span class="st">&#39;KCMSM&#39;</span>,</a>
<a class="sourceLine" id="cb5-109" title="109">                                        county<span class="op">==</span><span class="st">&#39;king&#39;</span><span class="op">&amp;</span>transm2<span class="op">!=</span><span class="st">&#39;MSM&#39;</span><span class="op">~</span><span class="st">&#39;KCnonMSM&#39;</span>,</a>
<a class="sourceLine" id="cb5-110" title="110">                                        county<span class="op">!=</span><span class="st">&#39;king&#39;</span><span class="op">~</span><span class="st">&#39;WAState&#39;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-111" title="111"><span class="st">            </span><span class="kw">mutate</span>(<span class="dt">TAXA_NEW:=</span> <span class="kw">paste0</span>(WRLD,<span class="st">&#39;___&#39;</span>,TAXA,<span class="st">&#39;_&#39;</span>,newnum,<span class="st">&#39;_&#39;</span>,county,<span class="st">&#39;_&#39;</span>,race2,<span class="st">&#39;_&#39;</span>,birthCountry2,<span class="st">&#39;_&#39;</span>,Gender2,<span class="st">&#39;_&#39;</span>,transm2,<span class="st">&#39;_&#39;</span>,seqy)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-112" title="112"><span class="st">            </span><span class="kw">select</span>(ST,WRLD,TAXA,TAXA_NEW)</a>
<a class="sourceLine" id="cb5-113" title="113">    tmp &lt;-<span class="st"> </span>dl <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">left_join</span>(dco, <span class="dt">by=</span><span class="st">&#39;CNTRY&#39;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-114" title="114"><span class="st">            </span><span class="kw">mutate</span>(<span class="dt">TAXA_NEW:=</span> <span class="kw">paste0</span>(WRLD,<span class="st">&#39;___&#39;</span>,TAXA)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-115" title="115"><span class="st">            </span><span class="kw">select</span>(ST,WRLD,TAXA,TAXA_NEW) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-116" title="116"><span class="st">            </span><span class="kw">rbind</span>(tmp) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-117" title="117"><span class="st">            </span><span class="kw">arrange</span>(WRLD)</a>
<a class="sourceLine" id="cb5-118" title="118">    outfile &lt;-<span class="st"> </span><span class="kw">paste0</span>(outfile.base,<span class="st">&#39;sequence_labels_KCMSM.csv&#39;</span>)</a>
<a class="sourceLine" id="cb5-119" title="119">    <span class="kw">write.csv</span>(tmp, <span class="dt">file=</span>outfile)</a>
<a class="sourceLine" id="cb5-120" title="120">    <span class="co">#</span></a>
<a class="sourceLine" id="cb5-121" title="121">    <span class="co">#   make sequence labels for KCHSX</span></a>
<a class="sourceLine" id="cb5-122" title="122">    <span class="co">#</span></a>
<a class="sourceLine" id="cb5-123" title="123">    tmp &lt;-<span class="st"> </span>dseq <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rename</span>(<span class="dt">TAXA:=</span> seqID) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-124" title="124"><span class="st">            </span><span class="kw">mutate</span>(<span class="dt">WRLD:=</span> <span class="kw">case_when</span>(county<span class="op">==</span><span class="st">&#39;king&#39;</span><span class="op">&amp;</span>transm2<span class="op">==</span><span class="st">&#39;HSX&#39;</span><span class="op">~</span><span class="st">&#39;KCHSX&#39;</span>,</a>
<a class="sourceLine" id="cb5-125" title="125">                            county<span class="op">==</span><span class="st">&#39;king&#39;</span><span class="op">&amp;</span>transm2<span class="op">!=</span><span class="st">&#39;HSX&#39;</span><span class="op">~</span><span class="st">&#39;KCnonHSX&#39;</span>,</a>
<a class="sourceLine" id="cb5-126" title="126">                            county<span class="op">!=</span><span class="st">&#39;king&#39;</span><span class="op">~</span><span class="st">&#39;WAState&#39;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-127" title="127"><span class="st">            </span><span class="kw">mutate</span>(<span class="dt">TAXA_NEW:=</span> <span class="kw">paste0</span>(WRLD,<span class="st">&#39;___&#39;</span>,TAXA,<span class="st">&#39;_&#39;</span>,newnum,<span class="st">&#39;_&#39;</span>,county,<span class="st">&#39;_&#39;</span>,race2,<span class="st">&#39;_&#39;</span>,birthCountry2,<span class="st">&#39;_&#39;</span>,Gender2,<span class="st">&#39;_&#39;</span>,transm2,<span class="st">&#39;_&#39;</span>,seqy)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-128" title="128"><span class="st">            </span><span class="kw">select</span>(ST,WRLD,TAXA,TAXA_NEW)</a>
<a class="sourceLine" id="cb5-129" title="129">    tmp &lt;-<span class="st"> </span>dl <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">left_join</span>(dco, <span class="dt">by=</span><span class="st">&#39;CNTRY&#39;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-130" title="130"><span class="st">            </span><span class="kw">mutate</span>(<span class="dt">TAXA_NEW:=</span> <span class="kw">paste0</span>(WRLD,<span class="st">&#39;___&#39;</span>,TAXA)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-131" title="131"><span class="st">            </span><span class="kw">select</span>(ST,WRLD,TAXA,TAXA_NEW) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-132" title="132"><span class="st">            </span><span class="kw">rbind</span>(tmp) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-133" title="133"><span class="st">            </span><span class="kw">arrange</span>(WRLD)</a>
<a class="sourceLine" id="cb5-134" title="134">    outfile &lt;-<span class="st"> </span><span class="kw">paste0</span>(outfile.base,<span class="st">&#39;sequence_labels_KCHSX.csv&#39;</span>)</a>
<a class="sourceLine" id="cb5-135" title="135">    <span class="kw">write.csv</span>(tmp, <span class="dt">file=</span>outfile)</a>
<a class="sourceLine" id="cb5-136" title="136">    <span class="co">#</span></a>
<a class="sourceLine" id="cb5-137" title="137">    <span class="co">#   make sequence labels for KC overall</span></a>
<a class="sourceLine" id="cb5-138" title="138">    <span class="co">#</span></a>
<a class="sourceLine" id="cb5-139" title="139">    tmp &lt;-<span class="st"> </span>dseq <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rename</span>(<span class="dt">TAXA:=</span> seqID) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-140" title="140"><span class="st">            </span><span class="kw">mutate</span>(<span class="dt">WRLD:=</span> <span class="kw">case_when</span>(county<span class="op">==</span><span class="st">&#39;king&#39;</span><span class="op">~</span><span class="st">&#39;KC&#39;</span>,                            </a>
<a class="sourceLine" id="cb5-141" title="141">                            county<span class="op">!=</span><span class="st">&#39;king&#39;</span><span class="op">~</span><span class="st">&#39;WAState&#39;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-142" title="142"><span class="st">            </span><span class="kw">mutate</span>(<span class="dt">TAXA_NEW:=</span> <span class="kw">paste0</span>(WRLD,<span class="st">&#39;___&#39;</span>,TAXA,<span class="st">&#39;_&#39;</span>,newnum,<span class="st">&#39;_&#39;</span>,county,<span class="st">&#39;_&#39;</span>,race2,<span class="st">&#39;_&#39;</span>,birthCountry2,<span class="st">&#39;_&#39;</span>,Gender2,<span class="st">&#39;_&#39;</span>,transm2,<span class="st">&#39;_&#39;</span>,seqy)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-143" title="143"><span class="st">            </span><span class="kw">select</span>(ST,WRLD,TAXA,TAXA_NEW)</a>
<a class="sourceLine" id="cb5-144" title="144">    tmp &lt;-<span class="st"> </span>dl <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">left_join</span>(dco, <span class="dt">by=</span><span class="st">&#39;CNTRY&#39;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-145" title="145"><span class="st">            </span><span class="kw">mutate</span>(<span class="dt">TAXA_NEW:=</span> <span class="kw">paste0</span>(WRLD,<span class="st">&#39;___&#39;</span>,TAXA)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-146" title="146"><span class="st">            </span><span class="kw">select</span>(ST,WRLD,TAXA,TAXA_NEW) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-147" title="147"><span class="st">            </span><span class="kw">rbind</span>(tmp) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-148" title="148"><span class="st">            </span><span class="kw">arrange</span>(WRLD)</a>
<a class="sourceLine" id="cb5-149" title="149">    outfile &lt;-<span class="st"> </span><span class="kw">paste0</span>(outfile.base,<span class="st">&#39;sequence_labels_KC.csv&#39;</span>)</a>
<a class="sourceLine" id="cb5-150" title="150">    <span class="kw">write.csv</span>(tmp, <span class="dt">file=</span>outfile)</a>
<a class="sourceLine" id="cb5-151" title="151">}</a></code></pre></div>
<p>Second, here is the code to setup the phylogeographic ancestral state reconstruction. For each subtype B tree, the maximum-parsimony analysis takes a few days. For the other subtypes, the results are pretty much instant.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1">seattle.<span class="fl">191017.</span>phyloscanner.nonB &lt;-<span class="st"> </span><span class="cf">function</span>()</a>
<a class="sourceLine" id="cb6-2" title="2">{</a>
<a class="sourceLine" id="cb6-3" title="3">    <span class="kw">require</span>(data.table)</a>
<a class="sourceLine" id="cb6-4" title="4">    <span class="kw">require</span>(ape)</a>
<a class="sourceLine" id="cb6-5" title="5">    <span class="kw">require</span>(adephylo)</a>
<a class="sourceLine" id="cb6-6" title="6">    <span class="kw">require</span>(phytools)</a>
<a class="sourceLine" id="cb6-7" title="7">    </a>
<a class="sourceLine" id="cb6-8" title="8">    prog.phyloscanner_analyse_trees &lt;-<span class="st"> &#39;/rds/general/user/or105/home/libs_sandbox/phyloscanner/phyloscanner_analyse_trees.R&#39;</span></a>
<a class="sourceLine" id="cb6-9" title="9">    plot.phylogenies &lt;-<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb6-10" title="10">    home &lt;-<span class="st"> &#39;~/Box Sync/OR_Work/Seattle/analysis_191017&#39;</span></a>
<a class="sourceLine" id="cb6-11" title="11">    <span class="co">#home &lt;- &#39;/rds/general/project/ratmann_seattle_data_analysis/live/olli_191017&#39;</span></a>
<a class="sourceLine" id="cb6-12" title="12">    </a>
<a class="sourceLine" id="cb6-13" title="13">    indir.trees &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;fasttree&#39;</span>)</a>
<a class="sourceLine" id="cb6-14" title="14">    infile.labels &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">FLABEL=</span> <span class="kw">c</span>(  <span class="kw">file.path</span>(home,<span class="st">&#39;misc&#39;</span>,<span class="st">&#39;180709_sequence_labels_KC.csv&#39;</span>),</a>
<a class="sourceLine" id="cb6-15" title="15">                                            <span class="kw">file.path</span>(home,<span class="st">&#39;misc&#39;</span>,<span class="st">&#39;180709_sequence_labels_KCMSM.csv&#39;</span>),</a>
<a class="sourceLine" id="cb6-16" title="16">                                            <span class="kw">file.path</span>(home,<span class="st">&#39;misc&#39;</span>,<span class="st">&#39;180709_sequence_labels_KCHSX.csv&#39;</span>))</a>
<a class="sourceLine" id="cb6-17" title="17">                                )</a>
<a class="sourceLine" id="cb6-18" title="18">    outdir &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;phyloscanner&#39;</span>)</a>
<a class="sourceLine" id="cb6-19" title="19">    </a>
<a class="sourceLine" id="cb6-20" title="20">    </a>
<a class="sourceLine" id="cb6-21" title="21">    infiles &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">FIN=</span><span class="kw">list.files</span>(indir.trees, <span class="dt">pattern=</span><span class="st">&#39;</span><span class="ch">\\</span><span class="st">.newick$&#39;</span>,<span class="dt">full.names=</span><span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb6-22" title="22">    infiles[, DUMMY<span class="op">:</span><span class="er">=</span><span class="st"> </span>1L]</a>
<a class="sourceLine" id="cb6-23" title="23">    infile.labels[, DUMMY<span class="op">:</span><span class="er">=</span><span class="st"> </span>1L]</a>
<a class="sourceLine" id="cb6-24" title="24">    infiles &lt;-<span class="st"> </span><span class="kw">merge</span>(infiles, infile.labels, <span class="dt">by=</span><span class="st">&#39;DUMMY&#39;</span>, <span class="dt">allow.cartesian=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb6-25" title="25">    infiles[, DUMMY<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="ot">NULL</span>]</a>
<a class="sourceLine" id="cb6-26" title="26">    infiles[, ST<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;.*_Subtype([A-Z0-9]+)_.*&#39;</span>,<span class="st">&#39;</span><span class="ch">\\</span><span class="st">1&#39;</span>,<span class="kw">basename</span>(FIN))]    </a>
<a class="sourceLine" id="cb6-27" title="27">    <span class="co">#infiles &lt;- subset(infiles, ST==&#39;B&#39;)</span></a>
<a class="sourceLine" id="cb6-28" title="28">    cmds &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&#39;list&#39;</span>,<span class="kw">nrow</span>(infiles))</a>
<a class="sourceLine" id="cb6-29" title="29">    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_len</span>(<span class="kw">nrow</span>(infiles)))</a>
<a class="sourceLine" id="cb6-30" title="30">    {</a>
<a class="sourceLine" id="cb6-31" title="31">        <span class="co">#i&lt;- 1</span></a>
<a class="sourceLine" id="cb6-32" title="32">        <span class="kw">cat</span>(<span class="st">&#39;</span><span class="ch">\n</span><span class="st">process &#39;</span>,infiles[i,FIN],<span class="st">&#39;</span><span class="ch">\n</span><span class="st">process &#39;</span>,infiles[i,FLABEL])</a>
<a class="sourceLine" id="cb6-33" title="33">        infile &lt;-<span class="st"> </span>infiles[i,FIN]</a>
<a class="sourceLine" id="cb6-34" title="34">        ph &lt;-<span class="st"> </span><span class="kw">read.tree</span>(infile)</a>
<a class="sourceLine" id="cb6-35" title="35">        ph<span class="op">$</span>node.label &lt;-<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb6-36" title="36">        <span class="co">#   update tip labels: add world region to start of label</span></a>
<a class="sourceLine" id="cb6-37" title="37">        local.world.region &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;^.*_labels_([A-Z]+)</span><span class="ch">\\</span><span class="st">.csv$&#39;</span>,<span class="st">&#39;</span><span class="ch">\\</span><span class="st">1&#39;</span>, <span class="kw">basename</span>(infiles[i,FLABEL]))</a>
<a class="sourceLine" id="cb6-38" title="38">        dl &lt;-<span class="st"> </span><span class="kw">as.data.table</span>(<span class="kw">read.csv</span>(infiles[i,FLABEL], <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>))</a>
<a class="sourceLine" id="cb6-39" title="39">        dl[, X<span class="op">:</span><span class="er">=</span><span class="ot">NULL</span>]   </a>
<a class="sourceLine" id="cb6-40" title="40">        <span class="kw">stopifnot</span>(<span class="op">!</span><span class="kw">any</span>(ph<span class="op">$</span>tip.label<span class="op">==</span><span class="st">&#39;&#39;</span>))</a>
<a class="sourceLine" id="cb6-41" title="41">        dp &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">IDX=</span><span class="dv">1</span><span class="op">:</span><span class="kw">Ntip</span>(ph), <span class="dt">TAXA=</span>ph<span class="op">$</span>tip.label)</a>
<a class="sourceLine" id="cb6-42" title="42">        dp &lt;-<span class="st"> </span><span class="kw">merge</span>(dp,dl,<span class="dt">by=</span><span class="st">&#39;TAXA&#39;</span>,<span class="dt">all.x=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb6-43" title="43">        <span class="kw">stopifnot</span>( <span class="kw">nrow</span>(dp[<span class="kw">is.na</span>(TAXA_NEW),])<span class="op">==</span><span class="dv">0</span> )</a>
<a class="sourceLine" id="cb6-44" title="44">        <span class="kw">stopifnot</span>( <span class="op">!</span><span class="kw">any</span>(<span class="kw">duplicated</span>(ph<span class="op">$</span>tip.label)) )</a>
<a class="sourceLine" id="cb6-45" title="45">        ph<span class="op">$</span>tip.label &lt;-<span class="st"> </span>dp[<span class="kw">order</span>(IDX),][, TAXA_NEW]</a>
<a class="sourceLine" id="cb6-46" title="46">        <span class="co">#   re-root</span></a>
<a class="sourceLine" id="cb6-47" title="47">        tmp &lt;-<span class="st"> </span>dp[, <span class="kw">list</span>(<span class="dt">NST=</span><span class="kw">length</span>(TAXA)), by=<span class="st">&#39;ST&#39;</span>]</a>
<a class="sourceLine" id="cb6-48" title="48">        tmp &lt;-<span class="st"> </span>tmp[<span class="kw">order</span>(<span class="op">-</span>NST),][<span class="dv">2</span>,ST]</a>
<a class="sourceLine" id="cb6-49" title="49">        tmp &lt;-<span class="st"> </span><span class="kw">subset</span>(dp, ST<span class="op">==</span>tmp,)[,TAXA_NEW]</a>
<a class="sourceLine" id="cb6-50" title="50">        root &lt;-<span class="st"> </span><span class="kw">getMRCA</span>(ph, tmp)</a>
<a class="sourceLine" id="cb6-51" title="51">        ph &lt;-<span class="st"> </span><span class="kw">reroot</span>(ph, root, ph<span class="op">$</span>edge.length[<span class="kw">which</span>(ph<span class="op">$</span>edge[,<span class="dv">2</span>]<span class="op">==</span>root)]<span class="op">/</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb6-52" title="52">        <span class="co">#   drop other subtypes</span></a>
<a class="sourceLine" id="cb6-53" title="53">        tmp &lt;-<span class="st"> </span>dp[, <span class="kw">list</span>(<span class="dt">NST=</span><span class="kw">length</span>(TAXA)), by=<span class="st">&#39;ST&#39;</span>]</a>
<a class="sourceLine" id="cb6-54" title="54">        tmp &lt;-<span class="st"> </span>tmp[<span class="kw">order</span>(<span class="op">-</span>NST),][<span class="op">-</span><span class="dv">1</span>,ST]</a>
<a class="sourceLine" id="cb6-55" title="55">        tmp &lt;-<span class="st"> </span><span class="kw">subset</span>(dp, ST<span class="op">%in%</span>tmp,)[,TAXA_NEW]</a>
<a class="sourceLine" id="cb6-56" title="56">        ph &lt;-<span class="st"> </span><span class="kw">drop.tip</span>(ph, tmp)</a>
<a class="sourceLine" id="cb6-57" title="57">        <span class="co">#   write to file</span></a>
<a class="sourceLine" id="cb6-58" title="58">        intree.phsc &lt;-<span class="st"> </span><span class="kw">file.path</span>(outdir,<span class="kw">gsub</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">.newick&#39;</span>,<span class="kw">paste0</span>(<span class="st">&#39;_rerooted_&#39;</span>,local.world.region,<span class="st">&#39;.newick&#39;</span>),<span class="kw">basename</span>(infile)))</a>
<a class="sourceLine" id="cb6-59" title="59">        <span class="kw">write.tree</span>(ph, <span class="dt">file=</span>intree.phsc)</a>
<a class="sourceLine" id="cb6-60" title="60">        <span class="cf">if</span>(plot.phylogenies)</a>
<a class="sourceLine" id="cb6-61" title="61">        {</a>
<a class="sourceLine" id="cb6-62" title="62">            <span class="kw">pdf</span>(<span class="dt">file=</span><span class="kw">gsub</span>(<span class="st">&#39;newick&#39;</span>,<span class="st">&#39;pdf&#39;</span>,intree.phsc), <span class="dt">w=</span><span class="dv">20</span>, <span class="dt">h=</span><span class="dv">10</span><span class="op">+</span><span class="kw">Ntip</span>(ph)<span class="op">/</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb6-63" title="63">            <span class="kw">plot</span>(ph, <span class="dt">show.node.label=</span><span class="ot">TRUE</span>, <span class="dt">cex=</span><span class="fl">0.3</span>)</a>
<a class="sourceLine" id="cb6-64" title="64">            <span class="kw">dev.off</span>()           </a>
<a class="sourceLine" id="cb6-65" title="65">        }</a>
<a class="sourceLine" id="cb6-66" title="66">        <span class="co">#   make phyloscanner UNIX command</span></a>
<a class="sourceLine" id="cb6-67" title="67">        infile &lt;-<span class="st"> </span>intree.phsc</a>
<a class="sourceLine" id="cb6-68" title="68">        outputString &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="kw">gsub</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">.newick&#39;</span>,<span class="st">&#39;_&#39;</span>,intree.phsc))</a>
<a class="sourceLine" id="cb6-69" title="69">        tip.regex &lt;-<span class="st"> &quot;^([A-Za-z]+)___.*$&quot;</span></a>
<a class="sourceLine" id="cb6-70" title="70">        cmd &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;CWD=$(pwd)</span><span class="ch">\n</span><span class="st">&quot;</span>,<span class="dt">sep=</span><span class="st">&#39;&#39;</span>)</a>
<a class="sourceLine" id="cb6-71" title="71">        cmd &lt;-<span class="st"> </span><span class="kw">paste</span>(cmd,<span class="st">&quot;echo $CWD</span><span class="ch">\n</span><span class="st">&quot;</span>,<span class="dt">sep=</span><span class="st">&#39;&#39;</span>)  </a>
<a class="sourceLine" id="cb6-72" title="72">        tmpdir.prefix &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&#39;phsc_&#39;</span>,<span class="kw">format</span>(<span class="kw">Sys.time</span>(),<span class="st">&quot;%y-%m-%d-%H-%M-%S&quot;</span>),<span class="dt">sep=</span><span class="st">&#39;&#39;</span>)</a>
<a class="sourceLine" id="cb6-73" title="73">        tmpdir &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;$CWD/&quot;</span>,tmpdir.prefix,<span class="dt">sep=</span><span class="st">&#39;&#39;</span>)</a>
<a class="sourceLine" id="cb6-74" title="74">        tmp.in &lt;-<span class="st"> </span><span class="kw">file.path</span>(tmpdir, <span class="kw">basename</span>(infile))</a>
<a class="sourceLine" id="cb6-75" title="75">        cmd &lt;-<span class="st"> </span><span class="kw">paste</span>(cmd,<span class="st">&quot;mkdir -p &quot;</span>,tmpdir,<span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>,<span class="dt">sep=</span><span class="st">&#39;&#39;</span>)</a>
<a class="sourceLine" id="cb6-76" title="76">        cmd &lt;-<span class="st"> </span><span class="kw">paste</span>(cmd,<span class="st">&#39;cp &quot;&#39;</span>,infile,<span class="st">&#39;&quot; &#39;</span>,tmp.in,<span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>, <span class="dt">sep=</span><span class="st">&#39;&#39;</span>)</a>
<a class="sourceLine" id="cb6-77" title="77">        cmd &lt;-<span class="st"> </span><span class="kw">paste</span>(cmd,<span class="st">&#39;cd &#39;</span>, tmpdir,<span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>, <span class="dt">sep=</span><span class="st">&#39;&#39;</span>)</a>
<a class="sourceLine" id="cb6-78" title="78">        cmd &lt;-<span class="st"> </span><span class="kw">paste0</span>(cmd,<span class="st">&#39;Rscript &#39;</span>,prog.phyloscanner_analyse_trees,<span class="st">&#39; &#39;</span>,<span class="kw">basename</span>(infile),<span class="st">&#39; &#39;</span>,<span class="kw">basename</span>(outputString))   </a>
<a class="sourceLine" id="cb6-79" title="79">        cmd &lt;-<span class="st"> </span><span class="kw">paste0</span>(cmd,<span class="st">&#39; s,0 -m 1e-5 -x &quot;&#39;</span>,tip.regex,<span class="st">&#39;&quot; -v 1 -ow -rda</span><span class="ch">\n</span><span class="st">&#39;</span>)</a>
<a class="sourceLine" id="cb6-80" title="80">        cmd &lt;-<span class="st"> </span><span class="kw">paste0</span>(cmd,<span class="st">&#39;mv &#39;</span>,<span class="kw">basename</span>(outputString),<span class="st">&#39;* &#39;</span>,<span class="st">&#39;&quot;&#39;</span>,<span class="kw">dirname</span>(outputString),<span class="st">&#39;&quot;&#39;</span>,<span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>) </a>
<a class="sourceLine" id="cb6-81" title="81">        cmd &lt;-<span class="st"> </span><span class="kw">paste</span>(cmd, <span class="st">&quot;cd $CWD</span><span class="ch">\n</span><span class="st">&quot;</span>,<span class="dt">sep=</span><span class="st">&#39;&#39;</span>)</a>
<a class="sourceLine" id="cb6-82" title="82">        cmd &lt;-<span class="st"> </span><span class="kw">paste</span>(cmd, <span class="st">&quot;rm &quot;</span>, tmpdir,<span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>,<span class="dt">sep=</span><span class="st">&#39;&#39;</span>)</a>
<a class="sourceLine" id="cb6-83" title="83">        cmds[[i]] &lt;-<span class="st"> </span>cmd</a>
<a class="sourceLine" id="cb6-84" title="84">    }</a>
<a class="sourceLine" id="cb6-85" title="85">    infiles[, CMD<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">unlist</span>(cmds)]   </a>
<a class="sourceLine" id="cb6-86" title="86">    </a>
<a class="sourceLine" id="cb6-87" title="87">    <span class="co">#   submit jobs like this one:</span></a>
<a class="sourceLine" id="cb6-88" title="88">    <span class="kw">cat</span>(infiles[<span class="dv">1</span>,CMD])</a>
<a class="sourceLine" id="cb6-89" title="89">    </a>
<a class="sourceLine" id="cb6-90" title="90">    <span class="co">#   run on HPC as array job</span></a>
<a class="sourceLine" id="cb6-91" title="91">    df &lt;-<span class="st"> </span><span class="kw">copy</span>(infiles)</a>
<a class="sourceLine" id="cb6-92" title="92">    df[, CASE_ID<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(df)]  </a>
<a class="sourceLine" id="cb6-93" title="93">    <span class="co">#   make PBS header</span></a>
<a class="sourceLine" id="cb6-94" title="94">    hpc.load    &lt;-<span class="st"> &quot;module load anaconda3/personal&quot;</span></a>
<a class="sourceLine" id="cb6-95" title="95">    hpc.select  &lt;-<span class="st"> </span><span class="dv">1</span>                        <span class="co"># number of nodes</span></a>
<a class="sourceLine" id="cb6-96" title="96">    hpc.nproc   &lt;-<span class="st"> </span><span class="dv">1</span>                        <span class="co"># number of processors on node</span></a>
<a class="sourceLine" id="cb6-97" title="97">    hpc.walltime&lt;-<span class="st"> </span><span class="dv">923</span>                      <span class="co"># walltime</span></a>
<a class="sourceLine" id="cb6-98" title="98">    hpc.q       &lt;-<span class="st"> &quot;pqeelab&quot;</span>                        <span class="co"># PBS queue</span></a>
<a class="sourceLine" id="cb6-99" title="99">    hpc.mem     &lt;-<span class="st"> &quot;24gb&quot;</span>                   <span class="co"># RAM</span></a>
<a class="sourceLine" id="cb6-100" title="100">    hpc.array   &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(df<span class="op">$</span>CASE_ID))   <span class="co"># number of runs for job array</span></a>
<a class="sourceLine" id="cb6-101" title="101">    pbshead     &lt;-<span class="st"> &quot;#!/bin/sh&quot;</span></a>
<a class="sourceLine" id="cb6-102" title="102">    tmp         &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;#PBS -l walltime=&quot;</span>, hpc.walltime, <span class="st">&quot;:59:00,pcput=&quot;</span>, hpc.walltime, <span class="st">&quot;:45:00&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb6-103" title="103">    pbshead     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb6-104" title="104">    tmp         &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;#PBS -l select=&quot;</span>, hpc.select, <span class="st">&quot;:ncpus=&quot;</span>, hpc.nproc,<span class="st">&quot;:mem=&quot;</span>, hpc.mem, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb6-105" title="105">    pbshead     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb6-106" title="106">    pbshead     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="st">&quot;#PBS -j oe&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>) </a>
<a class="sourceLine" id="cb6-107" title="107">    <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.na</span>(hpc.array))</a>
<a class="sourceLine" id="cb6-108" title="108">        pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">#PBS -J 1-&quot;</span>, hpc.array, <span class="dt">sep=</span><span class="st">&#39;&#39;</span>)        </a>
<a class="sourceLine" id="cb6-109" title="109">    <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.na</span>(hpc.q)) </a>
<a class="sourceLine" id="cb6-110" title="110">        pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="kw">paste</span>(<span class="st">&quot;#PBS -q&quot;</span>, hpc.q), <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb6-111" title="111">    pbshead     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, hpc.load, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)         </a>
<a class="sourceLine" id="cb6-112" title="112">    <span class="co">#   make array job</span></a>
<a class="sourceLine" id="cb6-113" title="113">    cmd     &lt;-<span class="st"> </span>df[, <span class="kw">list</span>(<span class="dt">CASE=</span><span class="kw">paste0</span>(CASE_ID,<span class="st">&#39;)</span><span class="ch">\n</span><span class="st">&#39;</span>,CMD,<span class="st">&#39;;;</span><span class="ch">\n</span><span class="st">&#39;</span>)), by=<span class="st">&#39;CASE_ID&#39;</span>]</a>
<a class="sourceLine" id="cb6-114" title="114">    cmd     &lt;-<span class="st"> </span>cmd[, <span class="kw">paste0</span>(<span class="st">&#39;case $PBS_ARRAY_INDEX in</span><span class="ch">\n</span><span class="st">&#39;</span>,<span class="kw">paste0</span>(CASE, <span class="dt">collapse=</span><span class="st">&#39;&#39;</span>),<span class="st">&#39;esac&#39;</span>)]         </a>
<a class="sourceLine" id="cb6-115" title="115">    cmd     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead,cmd,<span class="dt">sep=</span><span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>)  </a>
<a class="sourceLine" id="cb6-116" title="116">    <span class="co">#   submit job</span></a>
<a class="sourceLine" id="cb6-117" title="117">    outfile     &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;:&#39;</span>,<span class="st">&#39;&#39;</span>,<span class="kw">paste</span>(<span class="st">&quot;phs&quot;</span>,<span class="kw">paste</span>(<span class="kw">strsplit</span>(<span class="kw">date</span>(),<span class="dt">split=</span><span class="st">&#39; &#39;</span>)[[<span class="dv">1</span>]],<span class="dt">collapse=</span><span class="st">&#39;_&#39;</span>,<span class="dt">sep=</span><span class="st">&#39;&#39;</span>),<span class="st">&#39;sh&#39;</span>,<span class="dt">sep=</span><span class="st">&#39;.&#39;</span>))</a>
<a class="sourceLine" id="cb6-118" title="118">    outfile     &lt;-<span class="st"> </span><span class="kw">file.path</span>(outdir, outfile)</a>
<a class="sourceLine" id="cb6-119" title="119">    <span class="kw">cat</span>(cmd, <span class="dt">file=</span>outfile)</a>
<a class="sourceLine" id="cb6-120" title="120">    cmd         &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;qsub&quot;</span>, outfile)</a>
<a class="sourceLine" id="cb6-121" title="121">    <span class="kw">cat</span>(cmd)</a>
<a class="sourceLine" id="cb6-122" title="122">    <span class="kw">cat</span>(<span class="kw">system</span>(cmd, <span class="dt">intern=</span> <span class="ot">TRUE</span>))  </a>
<a class="sourceLine" id="cb6-123" title="123">}</a></code></pre></div>
</div>
<div id="step-5-date-trees-with-ancestral-state-reconstruction" class="section level2">
<h2><strong>Step 5 (date trees with ancestral state reconstruction)</strong></h2>
<p>Most phylodynamic analyses use as input a dated phylogeny, so we also have to date the trees obtained in step 4. Again, there are several available programs to do this, and I opted for <a href="https://github.com/emvolz/treedater">tree.dater</a>.</p>
<p>I chose to date the tree after ancestral states are reconstructed. This is because the sampling locations of the LANL sequences provide information on the ancestral states, but we don t have sampling dates for all LANL sequences, and we need sampling dates for dating the tree. To exercise caution, I therefore reconstruct ancestral states first. Then, I drop LANL sequences without sampling dates from the tree, which does not change the states that are attributed to the remaining viral lineages. I found several instances in which the actual parental clade of a King County subgraph was lost due to missing sampling dates, so I believe doing things this way round is a bit more accurate. The final step involves dating the pruned tree with tree.dater:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">seattle.<span class="fl">191017.</span>treedater.run &lt;-<span class="st"> </span><span class="cf">function</span>()</a>
<a class="sourceLine" id="cb7-2" title="2">{</a>
<a class="sourceLine" id="cb7-3" title="3">    <span class="kw">require</span>(big.phylo)</a>
<a class="sourceLine" id="cb7-4" title="4">    <span class="kw">require</span>(data.table)</a>
<a class="sourceLine" id="cb7-5" title="5">    <span class="kw">require</span>(tidyverse)</a>
<a class="sourceLine" id="cb7-6" title="6">    <span class="kw">require</span>(ape)</a>
<a class="sourceLine" id="cb7-7" title="7">    <span class="kw">require</span>(phytools)</a>
<a class="sourceLine" id="cb7-8" title="8">    <span class="kw">require</span>(treedater)</a>
<a class="sourceLine" id="cb7-9" title="9">    <span class="kw">require</span>(phyloscannerR)</a>
<a class="sourceLine" id="cb7-10" title="10">    </a>
<a class="sourceLine" id="cb7-11" title="11">    home &lt;-<span class="st"> &#39;~/Box Sync/OR_Work/Seattle/analysis_191017&#39;</span></a>
<a class="sourceLine" id="cb7-12" title="12">    home &lt;-<span class="st"> &#39;/rds/general/project/ratmann_seattle_data_analysis/live/olli_191017&#39;</span></a>
<a class="sourceLine" id="cb7-13" title="13">    infile.seqinfo &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">dirname</span>(home),<span class="st">&#39;PHSKC-2018-07-09&#39;</span>,<span class="st">&#39;sequences_meta.rds&#39;</span>)  </a>
<a class="sourceLine" id="cb7-14" title="14">    indir.phsc  &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;phyloscanner&#39;</span>)</a>
<a class="sourceLine" id="cb7-15" title="15">    outdir &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;phyloscanner_dated&#39;</span>)</a>
<a class="sourceLine" id="cb7-16" title="16">    infiles.phsc &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">F=</span><span class="kw">list.files</span>(indir.phsc, <span class="dt">pattern=</span><span class="st">&#39;workspace.rda$&#39;</span>, <span class="dt">full.names=</span><span class="ot">TRUE</span>, <span class="dt">recursive=</span><span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb7-17" title="17">    <span class="co">#infiles.phsc &lt;- subset(infiles.phsc, grepl(&#39;000&#39;,F))</span></a>
<a class="sourceLine" id="cb7-18" title="18">    alignment.length &lt;-<span class="st"> </span><span class="dv">1000</span></a>
<a class="sourceLine" id="cb7-19" title="19">    </a>
<a class="sourceLine" id="cb7-20" title="20">    <span class="co">#</span></a>
<a class="sourceLine" id="cb7-21" title="21">    <span class="co"># read Seattle sampling data </span></a>
<a class="sourceLine" id="cb7-22" title="22">    <span class="co">#</span></a>
<a class="sourceLine" id="cb7-23" title="23">    </a>
<a class="sourceLine" id="cb7-24" title="24">    dseq &lt;-<span class="st"> </span><span class="kw">readRDS</span>(infile.seqinfo) </a>
<a class="sourceLine" id="cb7-25" title="25">    dseq &lt;-<span class="st"> </span>dseq <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb7-26" title="26"><span class="st">            </span><span class="kw">select</span>(seqID, newnum, type, seqy, seqm) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb7-27" title="27"><span class="st">            </span><span class="kw">mutate</span>( <span class="dt">SEQ_DATE:=</span> seqy <span class="op">+</span><span class="st"> </span>(seqm<span class="dv">-1</span>)<span class="op">/</span><span class="dv">12</span> <span class="op">+</span><span class="st"> </span><span class="dv">15</span><span class="op">/</span><span class="dv">365</span>,</a>
<a class="sourceLine" id="cb7-28" title="28">                    <span class="dt">SEQ_DATE_LOWER:=</span> seqy <span class="op">+</span><span class="st"> </span>(seqm<span class="dv">-1</span>)<span class="op">/</span><span class="dv">12</span>,</a>
<a class="sourceLine" id="cb7-29" title="29">                    <span class="dt">SEQ_DATE_UPPER:=</span> seqy <span class="op">+</span><span class="st"> </span>(seqm<span class="dv">-1</span>)<span class="op">/</span><span class="dv">12</span> <span class="op">+</span><span class="st"> </span><span class="dv">30</span><span class="op">/</span><span class="dv">365</span></a>
<a class="sourceLine" id="cb7-30" title="30">                    ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb7-31" title="31"><span class="st">            </span><span class="kw">select</span>(<span class="op">-</span>seqy, <span class="op">-</span>seqm) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb7-32" title="32"><span class="st">            </span><span class="kw">rename</span>(<span class="dt">TAXA=</span> seqID, <span class="dt">PID=</span> newnum, <span class="dt">TYPE=</span>type)</a>
<a class="sourceLine" id="cb7-33" title="33">    dseq &lt;-<span class="st"> </span><span class="kw">as.data.table</span>(dseq)</a>
<a class="sourceLine" id="cb7-34" title="34">    </a>
<a class="sourceLine" id="cb7-35" title="35">    <span class="co">#</span></a>
<a class="sourceLine" id="cb7-36" title="36">    <span class="co">#   for each tree: </span></a>
<a class="sourceLine" id="cb7-37" title="37">    <span class="co">#   make data.table of sequence sampling times </span></a>
<a class="sourceLine" id="cb7-38" title="38">    <span class="co">#   remove taxa without data on sampling times</span></a>
<a class="sourceLine" id="cb7-39" title="39">    <span class="co">#</span></a>
<a class="sourceLine" id="cb7-40" title="40">    cmds &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&#39;list&#39;</span>,<span class="kw">nrow</span>(infiles.phsc))</a>
<a class="sourceLine" id="cb7-41" title="41">    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_len</span>(<span class="kw">nrow</span>(infiles.phsc)))</a>
<a class="sourceLine" id="cb7-42" title="42">    {</a>
<a class="sourceLine" id="cb7-43" title="43">        <span class="co">#   i&lt;- 4</span></a>
<a class="sourceLine" id="cb7-44" title="44">        <span class="kw">cat</span>(<span class="st">&#39;</span><span class="ch">\n</span><span class="st">Process&#39;</span>,i)</a>
<a class="sourceLine" id="cb7-45" title="45">        infile &lt;-<span class="st"> </span>infiles.phsc[i,F] </a>
<a class="sourceLine" id="cb7-46" title="46">        <span class="kw">load</span>(infile)</a>
<a class="sourceLine" id="cb7-47" title="47">        ph &lt;-<span class="st"> </span>phyloscanner.trees[[<span class="dv">1</span>]][[<span class="st">&#39;tree&#39;</span>]]</a>
<a class="sourceLine" id="cb7-48" title="48">        <span class="kw">stopifnot</span>( <span class="op">!</span><span class="kw">any</span>( ph<span class="op">$</span>tip.label<span class="op">==</span><span class="st">&#39;&#39;</span> ) )       </a>
<a class="sourceLine" id="cb7-49" title="49">        </a>
<a class="sourceLine" id="cb7-50" title="50">        <span class="co">#</span></a>
<a class="sourceLine" id="cb7-51" title="51">        <span class="co">#   drop tips without sequence dates, </span></a>
<a class="sourceLine" id="cb7-52" title="52">        <span class="co">#   while conserving the ancestral state reconstructions</span></a>
<a class="sourceLine" id="cb7-53" title="53">        <span class="co">#</span></a>
<a class="sourceLine" id="cb7-54" title="54">        </a>
<a class="sourceLine" id="cb7-55" title="55">        <span class="co">#   extract taxa names</span></a>
<a class="sourceLine" id="cb7-56" title="56">        dph &lt;-<span class="st"> </span><span class="kw">data.table</span>(  <span class="dt">TAXA_LABEL=</span>ph<span class="op">$</span>tip.label,</a>
<a class="sourceLine" id="cb7-57" title="57">                <span class="dt">TAXA=</span> <span class="kw">gsub</span>(<span class="st">&#39;^(PR/RT-[0-9]+).*&#39;</span>,<span class="st">&#39;</span><span class="ch">\\</span><span class="st">1&#39;</span>,<span class="kw">gsub</span>(<span class="st">&#39;^[^_]+___(.*)&#39;</span>,<span class="st">&#39;</span><span class="ch">\\</span><span class="st">1&#39;</span>,ph<span class="op">$</span>tip.label)),</a>
<a class="sourceLine" id="cb7-58" title="58">                <span class="dt">TAXA_ID=</span> <span class="kw">seq_along</span>(ph<span class="op">$</span>tip.label))</a>
<a class="sourceLine" id="cb7-59" title="59">        <span class="co">#   add Seattle sequence dates</span></a>
<a class="sourceLine" id="cb7-60" title="60">        dph &lt;-<span class="st"> </span><span class="kw">merge</span>(dph, dseq, <span class="dt">by=</span><span class="st">&#39;TAXA&#39;</span>, <span class="dt">all.x=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb7-61" title="61">        <span class="co">#   extract GenBank sequence dates from taxa names where possible </span></a>
<a class="sourceLine" id="cb7-62" title="62">        dph[, GENBANK_ID<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;^([^</span><span class="ch">\\</span><span class="st">.]+)</span><span class="ch">\\</span><span class="st">.([^</span><span class="ch">\\</span><span class="st">.]+)</span><span class="ch">\\</span><span class="st">.([^</span><span class="ch">\\</span><span class="st">.]+)</span><span class="ch">\\</span><span class="st">.([^</span><span class="ch">\\</span><span class="st">.]+)</span><span class="ch">\\</span><span class="st">.([^</span><span class="ch">\\</span><span class="st">.]+)</span><span class="ch">\\</span><span class="st">.([^</span><span class="ch">\\</span><span class="st">.]+)$&#39;</span>,<span class="st">&#39;</span><span class="ch">\\</span><span class="st">5&#39;</span>,TAXA)]</a>
<a class="sourceLine" id="cb7-63" title="63">        dph[, GENBANK_SEQDATE<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;^([^</span><span class="ch">\\</span><span class="st">.]+)</span><span class="ch">\\</span><span class="st">.([^</span><span class="ch">\\</span><span class="st">.]+)</span><span class="ch">\\</span><span class="st">.([^</span><span class="ch">\\</span><span class="st">.]+)</span><span class="ch">\\</span><span class="st">.([^</span><span class="ch">\\</span><span class="st">.]+)</span><span class="ch">\\</span><span class="st">.([^</span><span class="ch">\\</span><span class="st">.]+)</span><span class="ch">\\</span><span class="st">.([^</span><span class="ch">\\</span><span class="st">.]+)$&#39;</span>,<span class="st">&#39;</span><span class="ch">\\</span><span class="st">3&#39;</span>,TAXA)]</a>
<a class="sourceLine" id="cb7-64" title="64">        <span class="kw">set</span>(dph, dph[, <span class="kw">which</span>(GENBANK_SEQDATE<span class="op">==</span><span class="st">&#39;-&#39;</span> <span class="op">|</span><span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(SEQ_DATE))],<span class="st">&#39;GENBANK_SEQDATE&#39;</span>,<span class="ot">NA_character_</span>) </a>
<a class="sourceLine" id="cb7-65" title="65">        <span class="kw">set</span>(dph, <span class="ot">NULL</span>, <span class="st">&#39;GENBANK_SEQDATE_LOWER&#39;</span>, dph[, <span class="kw">as.numeric</span>(GENBANK_SEQDATE) ])</a>
<a class="sourceLine" id="cb7-66" title="66">        <span class="kw">set</span>(dph, <span class="ot">NULL</span>, <span class="st">&#39;GENBANK_SEQDATE_UPPER&#39;</span>, dph[, <span class="kw">as.numeric</span>(GENBANK_SEQDATE) <span class="op">+</span><span class="st"> </span><span class="dv">364</span><span class="op">/</span><span class="dv">365</span> ])  </a>
<a class="sourceLine" id="cb7-67" title="67">        <span class="kw">set</span>(dph, <span class="ot">NULL</span>, <span class="st">&#39;GENBANK_SEQDATE&#39;</span>, dph[, <span class="kw">as.numeric</span>(GENBANK_SEQDATE) <span class="op">+</span><span class="st"> </span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span> ])</a>
<a class="sourceLine" id="cb7-68" title="68">        tmp &lt;-<span class="st"> </span>dph[, <span class="kw">which</span>(<span class="kw">is.na</span>(SEQ_DATE))]</a>
<a class="sourceLine" id="cb7-69" title="69">        <span class="kw">set</span>(dph, tmp, <span class="st">&#39;SEQ_DATE&#39;</span>, dph[tmp, GENBANK_SEQDATE])</a>
<a class="sourceLine" id="cb7-70" title="70">        <span class="kw">set</span>(dph, tmp, <span class="st">&#39;SEQ_DATE_LOWER&#39;</span>, dph[tmp, GENBANK_SEQDATE_LOWER])</a>
<a class="sourceLine" id="cb7-71" title="71">        <span class="kw">set</span>(dph, tmp, <span class="st">&#39;SEQ_DATE_UPPER&#39;</span>, dph[tmp, GENBANK_SEQDATE_UPPER])</a>
<a class="sourceLine" id="cb7-72" title="72">        <span class="kw">set</span>(dph, <span class="ot">NULL</span>, <span class="kw">c</span>(<span class="st">&#39;GENBANK_SEQDATE&#39;</span>,<span class="st">&#39;GENBANK_SEQDATE_LOWER&#39;</span>,<span class="st">&#39;GENBANK_SEQDATE_UPPER&#39;</span>), <span class="ot">NULL</span>)  </a>
<a class="sourceLine" id="cb7-73" title="73">        <span class="co">#   drop tips </span></a>
<a class="sourceLine" id="cb7-74" title="74">        dph.old &lt;-<span class="st"> </span><span class="kw">subset</span>(dph, <span class="dt">select=</span><span class="kw">c</span>(TAXA, SEQ_DATE, SEQ_DATE_LOWER, SEQ_DATE_UPPER))    </a>
<a class="sourceLine" id="cb7-75" title="75">        tmp &lt;-<span class="st"> </span><span class="kw">subset</span>(dph, <span class="kw">is.na</span>(SEQ_DATE))[, TAXA_ID]</a>
<a class="sourceLine" id="cb7-76" title="76">        <span class="kw">cat</span>(<span class="st">&#39;</span><span class="ch">\n</span><span class="st">Dropping tips without sampling date from &#39;</span>, infile,<span class="st">&#39; n=&#39;</span>, <span class="kw">length</span>(tmp), <span class="st">&#39;of Ntips=&#39;</span>, <span class="kw">Ntip</span>(ph), <span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>)</a>
<a class="sourceLine" id="cb7-77" title="77">        ph &lt;-<span class="st"> </span><span class="kw">phyloscanner.to.simmap</span>(ph)    </a>
<a class="sourceLine" id="cb7-78" title="78">        ph &lt;-<span class="st"> </span>phytools<span class="op">:::</span><span class="kw">drop.tip.simmap</span>(ph, ph<span class="op">$</span>tip.label[tmp])</a>
<a class="sourceLine" id="cb7-79" title="79">        ph &lt;-<span class="st"> </span><span class="kw">simmap.to.phyloscanner</span>(ph)        </a>
<a class="sourceLine" id="cb7-80" title="80">        </a>
<a class="sourceLine" id="cb7-81" title="81">        <span class="co">#</span></a>
<a class="sourceLine" id="cb7-82" title="82">        <span class="co">#   date tree</span></a>
<a class="sourceLine" id="cb7-83" title="83">        <span class="co">#   </span></a>
<a class="sourceLine" id="cb7-84" title="84">        </a>
<a class="sourceLine" id="cb7-85" title="85">        <span class="co">#   make data.table of sequence sampling times</span></a>
<a class="sourceLine" id="cb7-86" title="86">        dph &lt;-<span class="st"> </span><span class="kw">data.table</span>(  <span class="dt">TAXA_LABEL=</span>ph<span class="op">$</span>tip.label,</a>
<a class="sourceLine" id="cb7-87" title="87">                <span class="dt">TAXA=</span> <span class="kw">gsub</span>(<span class="st">&#39;^(PR/RT-[0-9]+).*&#39;</span>,<span class="st">&#39;</span><span class="ch">\\</span><span class="st">1&#39;</span>,<span class="kw">gsub</span>(<span class="st">&#39;^[^_]+___(.*)&#39;</span>,<span class="st">&#39;</span><span class="ch">\\</span><span class="st">1&#39;</span>,ph<span class="op">$</span>tip.label)),</a>
<a class="sourceLine" id="cb7-88" title="88">                <span class="dt">TAXA_ID=</span> <span class="kw">seq_along</span>(ph<span class="op">$</span>tip.label))</a>
<a class="sourceLine" id="cb7-89" title="89">        dph &lt;-<span class="st"> </span><span class="kw">merge</span>(dph, dph.old, <span class="dt">by=</span> <span class="st">&#39;TAXA&#39;</span>)</a>
<a class="sourceLine" id="cb7-90" title="90">        <span class="kw">stopifnot</span>( <span class="op">!</span><span class="kw">any</span>(<span class="kw">is.na</span>(dph<span class="op">$</span>SEQ_DATE)) )</a>
<a class="sourceLine" id="cb7-91" title="91">        dph &lt;-<span class="st"> </span>dph[<span class="kw">order</span>(TAXA_ID),]</a>
<a class="sourceLine" id="cb7-92" title="92">        <span class="co">#   get into format needed for tree.dater   </span></a>
<a class="sourceLine" id="cb7-93" title="93">        sampling.times.init &lt;-<span class="st"> </span>dph<span class="op">$</span>SEQ_DATE</a>
<a class="sourceLine" id="cb7-94" title="94">        <span class="kw">names</span>(sampling.times.init) &lt;-<span class="st"> </span>dph<span class="op">$</span>TAXA_LABEL</a>
<a class="sourceLine" id="cb7-95" title="95">        sampling.times.bounds &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">subset</span>(dph, <span class="dt">select=</span><span class="kw">c</span>(SEQ_DATE_LOWER, SEQ_DATE_UPPER)))</a>
<a class="sourceLine" id="cb7-96" title="96">        <span class="kw">rownames</span>(sampling.times.bounds) &lt;-<span class="st"> </span>dph<span class="op">$</span>TAXA_LABEL       </a>
<a class="sourceLine" id="cb7-97" title="97">        <span class="kw">colnames</span>(sampling.times.bounds) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;lower&#39;</span>,<span class="st">&#39;upper&#39;</span>)</a>
<a class="sourceLine" id="cb7-98" title="98">        <span class="co">#   save files for dating</span></a>
<a class="sourceLine" id="cb7-99" title="99">        outfile.collapsed.phsc &lt;-<span class="st"> </span><span class="kw">file.path</span>(outdir, <span class="kw">gsub</span>(<span class="st">&#39;workspace</span><span class="ch">\\</span><span class="st">.rda&#39;</span>,<span class="st">&#39;collapsed_workspace</span><span class="ch">\\</span><span class="st">.rda&#39;</span>,<span class="kw">basename</span>(infile)))</a>
<a class="sourceLine" id="cb7-100" title="100">        outfile.tree &lt;-<span class="st"> </span><span class="kw">file.path</span>(outdir, <span class="kw">gsub</span>(<span class="st">&#39;workspace</span><span class="ch">\\</span><span class="st">.rda&#39;</span>,<span class="st">&#39;collapsed</span><span class="ch">\\</span><span class="st">.newick&#39;</span>,<span class="kw">basename</span>(infile)))</a>
<a class="sourceLine" id="cb7-101" title="101">        outfile.sampling.times.bounds &lt;-<span class="st"> </span><span class="kw">file.path</span>(outdir, <span class="kw">gsub</span>(<span class="st">&#39;workspace</span><span class="ch">\\</span><span class="st">.rda&#39;</span>,<span class="st">&#39;sampling_times_bounds</span><span class="ch">\\</span><span class="st">.csv&#39;</span>,<span class="kw">basename</span>(infile)))</a>
<a class="sourceLine" id="cb7-102" title="102">        outfile.sampling.times.init &lt;-<span class="st"> </span><span class="kw">file.path</span>(outdir, <span class="kw">gsub</span>(<span class="st">&#39;workspace</span><span class="ch">\\</span><span class="st">.rda&#39;</span>,<span class="st">&#39;sampling_times_init</span><span class="ch">\\</span><span class="st">.csv&#39;</span>,<span class="kw">basename</span>(infile)))</a>
<a class="sourceLine" id="cb7-103" title="103">        <span class="kw">save</span>(ph, <span class="dt">file=</span>outfile.collapsed.phsc)       </a>
<a class="sourceLine" id="cb7-104" title="104">        <span class="kw">write.tree</span>(ph, <span class="dt">file=</span>outfile.tree)</a>
<a class="sourceLine" id="cb7-105" title="105">        <span class="kw">write.csv</span>(sampling.times.bounds, <span class="dt">file=</span>outfile.sampling.times.bounds, <span class="dt">row.names=</span><span class="ot">TRUE</span>)        </a>
<a class="sourceLine" id="cb7-106" title="106">        <span class="kw">write.csv</span>(<span class="kw">data.table</span>(<span class="dt">TAXA=</span><span class="kw">names</span>(sampling.times.init), <span class="dt">SEQ_DATE=</span>sampling.times.init), <span class="dt">file=</span>outfile.sampling.times.init, <span class="dt">row.names=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb7-107" title="107">        control &lt;-<span class="st"> </span><span class="kw">list</span>( <span class="dt">outfile=</span><span class="kw">gsub</span>(<span class="st">&#39;.newick$&#39;</span>,<span class="st">&#39;_dated.newick&#39;</span>,outfile.tree), </a>
<a class="sourceLine" id="cb7-108" title="108">                         <span class="dt">ali.len=</span>alignment.length, </a>
<a class="sourceLine" id="cb7-109" title="109">                         <span class="dt">root=</span><span class="ot">NA</span>, </a>
<a class="sourceLine" id="cb7-110" title="110">                         <span class="dt">omega0=</span><span class="ot">NA</span>, </a>
<a class="sourceLine" id="cb7-111" title="111">                         <span class="dt">temporalConstraints=</span><span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb7-112" title="112">                         <span class="dt">strictClock=</span><span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb7-113" title="113">                         <span class="dt">estimateSampleTimes=</span>outfile.sampling.times.bounds </a>
<a class="sourceLine" id="cb7-114" title="114">                         )</a>
<a class="sourceLine" id="cb7-115" title="115">        cmd &lt;-<span class="st"> </span><span class="kw">cmd.treedater.script</span>(outfile.tree, outfile.sampling.times.init, <span class="dt">control=</span>control)</a>
<a class="sourceLine" id="cb7-116" title="116">        cmds[[i]] &lt;-<span class="st"> </span>cmd</a>
<a class="sourceLine" id="cb7-117" title="117">    }   </a>
<a class="sourceLine" id="cb7-118" title="118">    infiles.phsc[, CMD<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">unlist</span>(cmds)]  </a>
<a class="sourceLine" id="cb7-119" title="119">    </a>
<a class="sourceLine" id="cb7-120" title="120">    <span class="co">#   submit jobs like this one:</span></a>
<a class="sourceLine" id="cb7-121" title="121">    <span class="kw">cat</span>(infiles.phsc[<span class="dv">1</span>,CMD])</a>
<a class="sourceLine" id="cb7-122" title="122">    </a>
<a class="sourceLine" id="cb7-123" title="123">    <span class="co">#   run on HPC as array job</span></a>
<a class="sourceLine" id="cb7-124" title="124">    df &lt;-<span class="st"> </span><span class="kw">copy</span>(infiles.phsc)</a>
<a class="sourceLine" id="cb7-125" title="125">    df[, CASE_ID<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(df)]  </a>
<a class="sourceLine" id="cb7-126" title="126">    <span class="co">#   make PBS header</span></a>
<a class="sourceLine" id="cb7-127" title="127">    hpc.load    &lt;-<span class="st"> &quot;module load anaconda3/personal&quot;</span></a>
<a class="sourceLine" id="cb7-128" title="128">    hpc.select  &lt;-<span class="st"> </span><span class="dv">1</span>                        <span class="co"># number of nodes</span></a>
<a class="sourceLine" id="cb7-129" title="129">    hpc.nproc   &lt;-<span class="st"> </span><span class="dv">1</span>                        <span class="co"># number of processors on node</span></a>
<a class="sourceLine" id="cb7-130" title="130">    hpc.walltime&lt;-<span class="st"> </span><span class="dv">23</span>                       <span class="co"># walltime</span></a>
<a class="sourceLine" id="cb7-131" title="131">    hpc.q       &lt;-<span class="st"> </span><span class="ot">NA</span> <span class="co">#&quot;pqeelab&quot;                        # PBS queue</span></a>
<a class="sourceLine" id="cb7-132" title="132">    hpc.mem     &lt;-<span class="st"> &quot;4gb&quot;</span>                    <span class="co"># RAM</span></a>
<a class="sourceLine" id="cb7-133" title="133">    hpc.array   &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(df<span class="op">$</span>CASE_ID))   <span class="co"># number of runs for job array</span></a>
<a class="sourceLine" id="cb7-134" title="134">    pbshead     &lt;-<span class="st"> &quot;#!/bin/sh&quot;</span></a>
<a class="sourceLine" id="cb7-135" title="135">    tmp         &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;#PBS -l walltime=&quot;</span>, hpc.walltime, <span class="st">&quot;:59:00,pcput=&quot;</span>, hpc.walltime, <span class="st">&quot;:45:00&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb7-136" title="136">    pbshead     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb7-137" title="137">    tmp         &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;#PBS -l select=&quot;</span>, hpc.select, <span class="st">&quot;:ncpus=&quot;</span>, hpc.nproc,<span class="st">&quot;:mem=&quot;</span>, hpc.mem, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb7-138" title="138">    pbshead     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb7-139" title="139">    pbshead     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="st">&quot;#PBS -j oe&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>) </a>
<a class="sourceLine" id="cb7-140" title="140">    <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.na</span>(hpc.array))</a>
<a class="sourceLine" id="cb7-141" title="141">        pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">#PBS -J 1-&quot;</span>, hpc.array, <span class="dt">sep=</span><span class="st">&#39;&#39;</span>)        </a>
<a class="sourceLine" id="cb7-142" title="142">    <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.na</span>(hpc.q)) </a>
<a class="sourceLine" id="cb7-143" title="143">        pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="kw">paste</span>(<span class="st">&quot;#PBS -q&quot;</span>, hpc.q), <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb7-144" title="144">    pbshead     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, hpc.load, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)         </a>
<a class="sourceLine" id="cb7-145" title="145">    <span class="co">#   make array job</span></a>
<a class="sourceLine" id="cb7-146" title="146">    cmd     &lt;-<span class="st"> </span>df[, <span class="kw">list</span>(<span class="dt">CASE=</span><span class="kw">paste0</span>(CASE_ID,<span class="st">&#39;)</span><span class="ch">\n</span><span class="st">&#39;</span>,CMD,<span class="st">&#39;;;</span><span class="ch">\n</span><span class="st">&#39;</span>)), by=<span class="st">&#39;CASE_ID&#39;</span>]</a>
<a class="sourceLine" id="cb7-147" title="147">    cmd     &lt;-<span class="st"> </span>cmd[, <span class="kw">paste0</span>(<span class="st">&#39;case $PBS_ARRAY_INDEX in</span><span class="ch">\n</span><span class="st">&#39;</span>,<span class="kw">paste0</span>(CASE, <span class="dt">collapse=</span><span class="st">&#39;&#39;</span>),<span class="st">&#39;esac&#39;</span>)]         </a>
<a class="sourceLine" id="cb7-148" title="148">    cmd     &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead,cmd,<span class="dt">sep=</span><span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>)  </a>
<a class="sourceLine" id="cb7-149" title="149">    <span class="co">#   submit job</span></a>
<a class="sourceLine" id="cb7-150" title="150">    outfile     &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;:&#39;</span>,<span class="st">&#39;&#39;</span>,<span class="kw">paste</span>(<span class="st">&quot;phs&quot;</span>,<span class="kw">paste</span>(<span class="kw">strsplit</span>(<span class="kw">date</span>(),<span class="dt">split=</span><span class="st">&#39; &#39;</span>)[[<span class="dv">1</span>]],<span class="dt">collapse=</span><span class="st">&#39;_&#39;</span>,<span class="dt">sep=</span><span class="st">&#39;&#39;</span>),<span class="st">&#39;sh&#39;</span>,<span class="dt">sep=</span><span class="st">&#39;.&#39;</span>))</a>
<a class="sourceLine" id="cb7-151" title="151">    outfile     &lt;-<span class="st"> </span><span class="kw">file.path</span>(outdir, outfile)</a>
<a class="sourceLine" id="cb7-152" title="152">    <span class="kw">cat</span>(cmd, <span class="dt">file=</span>outfile)</a>
<a class="sourceLine" id="cb7-153" title="153">    cmd         &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;qsub&quot;</span>, outfile)</a>
<a class="sourceLine" id="cb7-154" title="154">    <span class="kw">cat</span>(cmd)</a>
<a class="sourceLine" id="cb7-155" title="155">    <span class="kw">cat</span>(<span class="kw">system</span>(cmd, <span class="dt">intern=</span> <span class="ot">TRUE</span>))  </a>
<a class="sourceLine" id="cb7-156" title="156">}   </a></code></pre></div>
<p>Tree.dater does not change tree topologies, so we can easily superimpose the dated branch lengths onto the tree with ancestral state reconstructions, and plot the result:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1">seattle.<span class="fl">191017.</span>treedater.postprocess &lt;-<span class="st"> </span><span class="cf">function</span>()</a>
<a class="sourceLine" id="cb8-2" title="2">{</a>
<a class="sourceLine" id="cb8-3" title="3">    <span class="kw">require</span>(data.table) </a>
<a class="sourceLine" id="cb8-4" title="4">    <span class="kw">require</span>(phyloscannerR)</a>
<a class="sourceLine" id="cb8-5" title="5">    </a>
<a class="sourceLine" id="cb8-6" title="6">    home &lt;-<span class="st"> &#39;~/Box Sync/OR_Work/Seattle/analysis_191017&#39;</span></a>
<a class="sourceLine" id="cb8-7" title="7">    <span class="co">#home &lt;- &#39;/rds/general/project/ratmann_seattle_data_analysis/live/olli_191017&#39;      </span></a>
<a class="sourceLine" id="cb8-8" title="8">    indir &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;phyloscanner_dated&#39;</span>)</a>
<a class="sourceLine" id="cb8-9" title="9">    infiles &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">F_PHSC=</span><span class="kw">list.files</span>(indir, <span class="dt">pattern=</span><span class="st">&#39;collapsed_workspace.rda$&#39;</span>, <span class="dt">full.names=</span><span class="ot">TRUE</span>, <span class="dt">recursive=</span><span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb8-10" title="10">    infiles[, BASENAME<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;collapsed_workspace.rda$&#39;</span>,<span class="st">&#39;&#39;</span>,<span class="kw">basename</span>(F_PHSC))]</a>
<a class="sourceLine" id="cb8-11" title="11">    tmp &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">F_TREE=</span><span class="kw">list.files</span>(indir, <span class="dt">pattern=</span><span class="st">&#39;collapsed_dated.newick$&#39;</span>, <span class="dt">full.names=</span><span class="ot">TRUE</span>, <span class="dt">recursive=</span><span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb8-12" title="12">    tmp[, BASENAME<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;collapsed_dated.newick$&#39;</span>,<span class="st">&#39;&#39;</span>,<span class="kw">basename</span>(F_TREE))]</a>
<a class="sourceLine" id="cb8-13" title="13">    infiles &lt;-<span class="st"> </span><span class="kw">merge</span>(infiles, tmp, <span class="dt">by=</span><span class="st">&#39;BASENAME&#39;</span>, <span class="dt">all.x=</span><span class="ot">TRUE</span>)   </a>
<a class="sourceLine" id="cb8-14" title="14">    <span class="kw">stopifnot</span>( <span class="kw">nrow</span>(<span class="kw">subset</span>(infiles, <span class="kw">is.na</span>(F_TREE)))<span class="op">==</span><span class="dv">0</span> )</a>
<a class="sourceLine" id="cb8-15" title="15">    </a>
<a class="sourceLine" id="cb8-16" title="16">    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_len</span>(<span class="kw">nrow</span>(infiles)))</a>
<a class="sourceLine" id="cb8-17" title="17">    {</a>
<a class="sourceLine" id="cb8-18" title="18">        i&lt;-<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb8-19" title="19">        <span class="kw">cat</span>(<span class="st">&#39;</span><span class="ch">\n</span><span class="st">Process &#39;</span>,infiles[i,F_TREE])</a>
<a class="sourceLine" id="cb8-20" title="20">        <span class="co">#   load data</span></a>
<a class="sourceLine" id="cb8-21" title="21">        <span class="kw">load</span>(infiles[i,F_PHSC])</a>
<a class="sourceLine" id="cb8-22" title="22">        ph.dated &lt;-<span class="st"> </span><span class="kw">read.tree</span>(infiles[i,F_TREE])        </a>
<a class="sourceLine" id="cb8-23" title="23">        <span class="kw">stopifnot</span>( <span class="kw">all</span>(  ph.dated<span class="op">$</span>tip.label <span class="op">==</span><span class="st"> </span>ph<span class="op">$</span>tip.label ) )</a>
<a class="sourceLine" id="cb8-24" title="24">        <span class="kw">stopifnot</span>( <span class="kw">all</span>( ph.dated<span class="op">$</span>edge <span class="op">==</span><span class="st"> </span>ph<span class="op">$</span>edge ) )</a>
<a class="sourceLine" id="cb8-25" title="25">        </a>
<a class="sourceLine" id="cb8-26" title="26">        <span class="co">#   since the tree topology is unchanged, we can copy</span></a>
<a class="sourceLine" id="cb8-27" title="27">        <span class="co">#   the branch lenghts in units of time onto the original tree</span></a>
<a class="sourceLine" id="cb8-28" title="28">        <span class="co">#   that has the ancestral state reconstructions</span></a>
<a class="sourceLine" id="cb8-29" title="29">        ph<span class="op">$</span>edge.length &lt;-<span class="st"> </span>ph.dated<span class="op">$</span>edge.length</a>
<a class="sourceLine" id="cb8-30" title="30">        </a>
<a class="sourceLine" id="cb8-31" title="31">        <span class="co">#   plot dated tree to spot obvious errors</span></a>
<a class="sourceLine" id="cb8-32" title="32">        outfile &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;collapsed_workspace</span><span class="ch">\\</span><span class="st">.rda&#39;</span>,<span class="st">&#39;annotated_dated_tree.pdf&#39;</span>,infiles[i,F_PHSC])</a>
<a class="sourceLine" id="cb8-33" title="33">        tmp &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&#39;list&#39;</span>)</a>
<a class="sourceLine" id="cb8-34" title="34">        tmp[[<span class="st">&#39;tree&#39;</span>]] &lt;-<span class="st"> </span>ph</a>
<a class="sourceLine" id="cb8-35" title="35">        tmp[[<span class="st">&#39;tree&#39;</span>]][[<span class="st">&#39;node.states&#39;</span>]] &lt;-<span class="st"> </span>tmp[[<span class="st">&#39;tree&#39;</span>]][[<span class="st">&#39;mapped.edge&#39;</span>]] &lt;-<span class="st"> </span>tmp[[<span class="st">&#39;tree&#39;</span>]][[<span class="st">&#39;maps&#39;</span>]] &lt;-<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb8-36" title="36">        <span class="kw">attr</span>(tmp[[<span class="st">&#39;tree&#39;</span>]],<span class="st">&#39;map.order&#39;</span>) &lt;-<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb8-37" title="37">        <span class="kw">attr</span>(tmp[[<span class="st">&#39;tree&#39;</span>]],<span class="st">&#39;class&#39;</span>) &lt;-<span class="st"> &#39;phylo&#39;</span></a>
<a class="sourceLine" id="cb8-38" title="38">        tmp[[<span class="st">&#39;read.counts&#39;</span>]] &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">1</span>, <span class="kw">Ntip</span>(ph))    </a>
<a class="sourceLine" id="cb8-39" title="39">        <span class="kw">write.annotated.tree</span>(tmp, outfile, <span class="dt">format=</span><span class="st">&quot;pdf&quot;</span>, <span class="dt">pdf.scale.bar.width =</span> <span class="fl">0.01</span>, <span class="dt">pdf.w =</span> <span class="dv">40</span>, <span class="dt">pdf.hm =</span> <span class="fl">0.2</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb8-40" title="40">        </a>
<a class="sourceLine" id="cb8-41" title="41">        <span class="co">#   save phyloscanner.tree</span></a>
<a class="sourceLine" id="cb8-42" title="42">        outfile &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;collapsed_workspace</span><span class="ch">\\</span><span class="st">.rda&#39;</span>,<span class="st">&#39;annotated_dated_tree.rda&#39;</span>,infiles[i,F_PHSC])</a>
<a class="sourceLine" id="cb8-43" title="43">        <span class="kw">save</span>(ph, <span class="dt">file=</span>outfile)</a>
<a class="sourceLine" id="cb8-44" title="44">    }   </a>
<a class="sourceLine" id="cb8-45" title="45">}</a></code></pre></div>
</div>
<div id="step-6-extract-phylogenetic-subgraphs-associated-with-viral-spread-in-king-county" class="section level2">
<h2><strong>Step 6 (extract phylogenetic subgraphs associated with viral spread in King County)</strong></h2>
<p>Finally, I extract the phylogenetic subgraphs that are associated with viral spread in King County/Seattle. My interpretation is that each subgraph corresponds to a distinct transmission chain in King County/Seattle, which may be partially observed due to incomplete sequence sampling.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">seattle.<span class="fl">191017.</span>phyloscanner.get.dated.subgraphs&lt;-<span class="st"> </span><span class="cf">function</span>()</a>
<a class="sourceLine" id="cb9-2" title="2">{</a>
<a class="sourceLine" id="cb9-3" title="3">    <span class="kw">require</span>(data.table)</a>
<a class="sourceLine" id="cb9-4" title="4">    <span class="kw">require</span>(phangorn)</a>
<a class="sourceLine" id="cb9-5" title="5">    <span class="kw">require</span>(ggplot2)</a>
<a class="sourceLine" id="cb9-6" title="6">    <span class="kw">require</span>(reshape)</a>
<a class="sourceLine" id="cb9-7" title="7">    <span class="kw">require</span>(phyloscannerR)</a>
<a class="sourceLine" id="cb9-8" title="8">    </a>
<a class="sourceLine" id="cb9-9" title="9">    <span class="co">#   working directory with phyloscanner output</span></a>
<a class="sourceLine" id="cb9-10" title="10">    home &lt;-<span class="st"> &#39;/Users/Oliver/Box Sync/OR_Work/Seattle/analysis_191017&#39;</span></a>
<a class="sourceLine" id="cb9-11" title="11">    indir.phsc  &lt;-<span class="st"> </span><span class="kw">file.path</span>(home,<span class="st">&#39;phyloscanner_dated&#39;</span>)</a>
<a class="sourceLine" id="cb9-12" title="12">    infiles     &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">F=</span><span class="kw">list.files</span>(indir.phsc, <span class="dt">pattern=</span><span class="st">&#39;_annotated_dated_tree.rda$&#39;</span>, <span class="dt">full.names=</span><span class="ot">TRUE</span>, <span class="dt">recursive=</span><span class="ot">TRUE</span>))  </a>
<a class="sourceLine" id="cb9-13" title="13">    infiles[, SELECT<span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;^.*_rerooted_([A-Za-z0-9]+)_.*$&#39;</span>,<span class="st">&#39;</span><span class="ch">\\</span><span class="st">1&#39;</span>,<span class="kw">basename</span>(F))]   </a>
<a class="sourceLine" id="cb9-14" title="14">    </a>
<a class="sourceLine" id="cb9-15" title="15">    <span class="co">#   extract subgraphs of dated trees</span></a>
<a class="sourceLine" id="cb9-16" title="16">    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_len</span>(<span class="kw">nrow</span>(infiles)))</a>
<a class="sourceLine" id="cb9-17" title="17">    {</a>
<a class="sourceLine" id="cb9-18" title="18">        <span class="co">#i&lt;- 21</span></a>
<a class="sourceLine" id="cb9-19" title="19">        <span class="kw">cat</span>(<span class="st">&#39;process&#39;</span>, i,<span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>)</a>
<a class="sourceLine" id="cb9-20" title="20">        infile &lt;-<span class="st"> </span>infiles[i, F]</a>
<a class="sourceLine" id="cb9-21" title="21">        host &lt;-<span class="st"> </span>infiles[i,SELECT]</a>
<a class="sourceLine" id="cb9-22" title="22">        <span class="kw">load</span>(infile)                    </a>
<a class="sourceLine" id="cb9-23" title="23">        mrcas &lt;-<span class="st"> </span><span class="kw">which</span>( <span class="kw">attr</span>(ph, <span class="st">&#39;SUBGRAPH_MRCA&#39;</span>) )</a>
<a class="sourceLine" id="cb9-24" title="24">        <span class="co">#   some of the tips states are &quot;unknown&quot; and this clashes internally with the NA state, so we need to take some extra care</span></a>
<a class="sourceLine" id="cb9-25" title="25">        <span class="co">#   this is not a problem because the &quot;unknown&quot; and NA state mean the same thing</span></a>
<a class="sourceLine" id="cb9-26" title="26">        <span class="kw">attr</span>(ph, <span class="st">&#39;INDIVIDUAL&#39;</span>) &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">attr</span>(ph, <span class="st">&#39;INDIVIDUAL&#39;</span>))</a>
<a class="sourceLine" id="cb9-27" title="27">        <span class="kw">attr</span>(ph, <span class="st">&#39;INDIVIDUAL&#39;</span>)[<span class="kw">is.na</span>(<span class="kw">attr</span>(ph, <span class="st">&#39;INDIVIDUAL&#39;</span>))] &lt;-<span class="st"> &#39;Unknown&#39;</span></a>
<a class="sourceLine" id="cb9-28" title="28">        mrcas &lt;-<span class="st"> </span>mrcas[ <span class="kw">attr</span>(ph, <span class="st">&#39;INDIVIDUAL&#39;</span>)[mrcas]<span class="op">==</span>host ]</a>
<a class="sourceLine" id="cb9-29" title="29">        <span class="kw">stopifnot</span>( <span class="op">!</span><span class="kw">any</span>(<span class="kw">is.na</span>(mrcas)) )     </a>
<a class="sourceLine" id="cb9-30" title="30">        <span class="co"># check that tree is of class simmap</span></a>
<a class="sourceLine" id="cb9-31" title="31">        <span class="kw">stopifnot</span>( <span class="kw">any</span>(<span class="kw">attr</span>(ph,<span class="st">&#39;class&#39;</span>)<span class="op">==</span><span class="st">&#39;simmap&#39;</span>) )</a>
<a class="sourceLine" id="cb9-32" title="32">        <span class="co"># extract subgraphs</span></a>
<a class="sourceLine" id="cb9-33" title="33">        subgraphs &lt;-<span class="st"> </span><span class="kw">lapply</span>(mrcas, <span class="cf">function</span>(mrca) <span class="kw">extract.subgraph</span>(ph, mrca))</a>
<a class="sourceLine" id="cb9-34" title="34">        <span class="co"># save</span></a>
<a class="sourceLine" id="cb9-35" title="35">        outfile &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;_annotated_dated_tree&#39;</span>,<span class="kw">paste0</span>(<span class="st">&#39;_datedsubgraphs_&#39;</span>,host),infile)</a>
<a class="sourceLine" id="cb9-36" title="36">        <span class="kw">save</span>(subgraphs, <span class="dt">file=</span>outfile)</a>
<a class="sourceLine" id="cb9-37" title="37">    }   </a>
<a class="sourceLine" id="cb9-38" title="38">}</a></code></pre></div>
<p>If we process the subgraphs, we observe that a large fraction of the subgraphs consist of just one sequence. These subgraphs don t contain information into viral spread within Seattle because no one else of the corresponding transmission chain was sampled, and so we can ignore them in a phylodynamic analysis. This reduces sample size, and makes the analysis more tractable.</p>
<p>On the other hand, the subgraphs of size 1 also provide useful information into the reproductive number, because they indicate stuttering viral spread, and a reproduction number below 1. So we don t want to exclude them when estimating reproductive numbers.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
